// aws-amplify/dist/cjs/utils/index.js
'use strict';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAmplifyConfig = exports.sharedInMemoryStorage = exports.sessionStorage = exports.defaultStorage = exports.CookieStorage = exports.ServiceWorker = exports.ConsoleLogger = exports.Cache = exports.I18n = exports.Hub = void 0;
/*
This file maps exports from `aws-amplify/utils`.
*/
var core_1 = require("@aws-amplify/core");
Object.defineProperty(exports, "Hub", { enumerable: true, get: function () { return core_1.Hub; } });
Object.defineProperty(exports, "I18n", { enumerable: true, get: function () { return core_1.I18n; } });
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return core_1.Cache; } });
Object.defineProperty(exports, "ConsoleLogger", { enumerable: true, get: function () { return core_1.ConsoleLogger; } });
Object.defineProperty(exports, "ServiceWorker", { enumerable: true, get: function () { return core_1.ServiceWorker; } });
Object.defineProperty(exports, "CookieStorage", { enumerable: true, get: function () { return core_1.CookieStorage; } });
Object.defineProperty(exports, "defaultStorage", { enumerable: true, get: function () { return core_1.defaultStorage; } });
Object.defineProperty(exports, "sessionStorage", { enumerable: true, get: function () { return core_1.sessionStorage; } });
Object.defineProperty(exports, "sharedInMemoryStorage", { enumerable: true, get: function () { return core_1.sharedInMemoryStorage; } });
var utils_1 = require("@aws-amplify/core/internals/utils");
Object.defineProperty(exports, "parseAmplifyConfig", { enumerable: true, get: function () { return utils_1.parseAWSExports; } });
