// @aws-amplify/storage/dist/cjs/providers/s3/utils/client/runtime/index.native.js
'use strict';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.utf8Encode = exports.toBase64 = exports.parser = exports.s3TransferHandler = exports.CONTENT_SHA256_HEADER = exports.CANCELED_ERROR_MESSAGE = exports.SEND_UPLOAD_PROGRESS_EVENT = exports.SEND_DOWNLOAD_PROGRESS_EVENT = void 0;
// Entry point for ReactNative-specific S3 client utilities
var constants_1 = require("./constants");
Object.defineProperty(exports, "SEND_DOWNLOAD_PROGRESS_EVENT", { enumerable: true, get: function () { return constants_1.SEND_DOWNLOAD_PROGRESS_EVENT; } });
Object.defineProperty(exports, "SEND_UPLOAD_PROGRESS_EVENT", { enumerable: true, get: function () { return constants_1.SEND_UPLOAD_PROGRESS_EVENT; } });
Object.defineProperty(exports, "CANCELED_ERROR_MESSAGE", { enumerable: true, get: function () { return constants_1.CANCELED_ERROR_MESSAGE; } });
Object.defineProperty(exports, "CONTENT_SHA256_HEADER", { enumerable: true, get: function () { return constants_1.CONTENT_SHA256_HEADER; } });
var xhr_1 = require("./s3TransferHandler/xhr");
Object.defineProperty(exports, "s3TransferHandler", { enumerable: true, get: function () { return xhr_1.s3TransferHandler; } });
var pureJs_1 = require("./xmlParser/pureJs");
Object.defineProperty(exports, "parser", { enumerable: true, get: function () { return pureJs_1.parser; } });
var index_native_1 = require("./base64/index.native");
Object.defineProperty(exports, "toBase64", { enumerable: true, get: function () { return index_native_1.toBase64; } });
Object.defineProperty(exports, "utf8Encode", { enumerable: true, get: function () { return index_native_1.utf8Encode; } });
