// @aws-amplify/core/dist/cjs/utils/index.js
'use strict';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.deDupeAsyncFunction = exports.deepFreeze = exports.urlSafeEncode = exports.urlSafeDecode = exports.retry = exports.jitteredExponentialRetry = exports.jitteredBackoff = exports.isNonRetryableError = exports.NonRetryableError = exports.isWebWorker = exports.isBrowser = exports.getClientInfo = exports.generateRandomString = void 0;
var generateRandomString_1 = require("./generateRandomString");
Object.defineProperty(exports, "generateRandomString", { enumerable: true, get: function () { return generateRandomString_1.generateRandomString; } });
var getClientInfo_1 = require("./getClientInfo");
Object.defineProperty(exports, "getClientInfo", { enumerable: true, get: function () { return getClientInfo_1.getClientInfo; } });
var isBrowser_1 = require("./isBrowser");
Object.defineProperty(exports, "isBrowser", { enumerable: true, get: function () { return isBrowser_1.isBrowser; } });
var isWebWorker_1 = require("./isWebWorker");
Object.defineProperty(exports, "isWebWorker", { enumerable: true, get: function () { return isWebWorker_1.isWebWorker; } });
var retry_1 = require("./retry");
Object.defineProperty(exports, "NonRetryableError", { enumerable: true, get: function () { return retry_1.NonRetryableError; } });
Object.defineProperty(exports, "isNonRetryableError", { enumerable: true, get: function () { return retry_1.isNonRetryableError; } });
Object.defineProperty(exports, "jitteredBackoff", { enumerable: true, get: function () { return retry_1.jitteredBackoff; } });
Object.defineProperty(exports, "jitteredExponentialRetry", { enumerable: true, get: function () { return retry_1.jitteredExponentialRetry; } });
Object.defineProperty(exports, "retry", { enumerable: true, get: function () { return retry_1.retry; } });
var urlSafeDecode_1 = require("./urlSafeDecode");
Object.defineProperty(exports, "urlSafeDecode", { enumerable: true, get: function () { return urlSafeDecode_1.urlSafeDecode; } });
var urlSafeEncode_1 = require("./urlSafeEncode");
Object.defineProperty(exports, "urlSafeEncode", { enumerable: true, get: function () { return urlSafeEncode_1.urlSafeEncode; } });
var deepFreeze_1 = require("./deepFreeze");
Object.defineProperty(exports, "deepFreeze", { enumerable: true, get: function () { return deepFreeze_1.deepFreeze; } });
var deDupeAsyncFunction_1 = require("./deDupeAsyncFunction");
Object.defineProperty(exports, "deDupeAsyncFunction", { enumerable: true, get: function () { return deDupeAsyncFunction_1.deDupeAsyncFunction; } });
