// @aws-sdk/client-bedrock-agent-runtime/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  AccessDeniedException: () => AccessDeniedException,
  ActionGroupInvocationInputFilterSensitiveLog: () => ActionGroupInvocationInputFilterSensitiveLog,
  ActionGroupInvocationOutputFilterSensitiveLog: () => ActionGroupInvocationOutputFilterSensitiveLog,
  BadGatewayException: () => BadGatewayException,
  BedrockAgentRuntime: () => BedrockAgentRuntime,
  BedrockAgentRuntimeClient: () => BedrockAgentRuntimeClient,
  BedrockAgentRuntimeServiceException: () => BedrockAgentRuntimeServiceException,
  ConflictException: () => ConflictException,
  CreationMode: () => CreationMode,
  DependencyFailedException: () => DependencyFailedException,
  FailureTraceFilterSensitiveLog: () => FailureTraceFilterSensitiveLog,
  FinalResponseFilterSensitiveLog: () => FinalResponseFilterSensitiveLog,
  InternalServerException: () => InternalServerException,
  InvocationInputFilterSensitiveLog: () => InvocationInputFilterSensitiveLog,
  InvocationType: () => InvocationType,
  InvokeAgentCommand: () => InvokeAgentCommand,
  InvokeAgentRequestFilterSensitiveLog: () => InvokeAgentRequestFilterSensitiveLog,
  InvokeAgentResponseFilterSensitiveLog: () => InvokeAgentResponseFilterSensitiveLog,
  KnowledgeBaseLookupInputFilterSensitiveLog: () => KnowledgeBaseLookupInputFilterSensitiveLog,
  KnowledgeBaseQueryFilterSensitiveLog: () => KnowledgeBaseQueryFilterSensitiveLog,
  ModelInvocationInputFilterSensitiveLog: () => ModelInvocationInputFilterSensitiveLog,
  ObservationFilterSensitiveLog: () => ObservationFilterSensitiveLog,
  OrchestrationTrace: () => OrchestrationTrace,
  OrchestrationTraceFilterSensitiveLog: () => OrchestrationTraceFilterSensitiveLog,
  PayloadPartFilterSensitiveLog: () => PayloadPartFilterSensitiveLog,
  PostProcessingModelInvocationOutputFilterSensitiveLog: () => PostProcessingModelInvocationOutputFilterSensitiveLog,
  PostProcessingParsedResponseFilterSensitiveLog: () => PostProcessingParsedResponseFilterSensitiveLog,
  PostProcessingTrace: () => PostProcessingTrace,
  PostProcessingTraceFilterSensitiveLog: () => PostProcessingTraceFilterSensitiveLog,
  PreProcessingModelInvocationOutputFilterSensitiveLog: () => PreProcessingModelInvocationOutputFilterSensitiveLog,
  PreProcessingParsedResponseFilterSensitiveLog: () => PreProcessingParsedResponseFilterSensitiveLog,
  PreProcessingTrace: () => PreProcessingTrace,
  PreProcessingTraceFilterSensitiveLog: () => PreProcessingTraceFilterSensitiveLog,
  PromptType: () => PromptType,
  RationaleFilterSensitiveLog: () => RationaleFilterSensitiveLog,
  RepromptResponseFilterSensitiveLog: () => RepromptResponseFilterSensitiveLog,
  ResourceNotFoundException: () => ResourceNotFoundException,
  ResponseStream: () => ResponseStream,
  ResponseStreamFilterSensitiveLog: () => ResponseStreamFilterSensitiveLog,
  RetrievalResultLocationType: () => RetrievalResultLocationType,
  RetrieveAndGenerateCommand: () => RetrieveAndGenerateCommand,
  RetrieveAndGenerateInputFilterSensitiveLog: () => RetrieveAndGenerateInputFilterSensitiveLog,
  RetrieveAndGenerateOutputFilterSensitiveLog: () => RetrieveAndGenerateOutputFilterSensitiveLog,
  RetrieveAndGenerateRequestFilterSensitiveLog: () => RetrieveAndGenerateRequestFilterSensitiveLog,
  RetrieveAndGenerateResponseFilterSensitiveLog: () => RetrieveAndGenerateResponseFilterSensitiveLog,
  RetrieveAndGenerateType: () => RetrieveAndGenerateType,
  RetrieveCommand: () => RetrieveCommand,
  RetrieveRequestFilterSensitiveLog: () => RetrieveRequestFilterSensitiveLog,
  RetrieveResponseFilterSensitiveLog: () => RetrieveResponseFilterSensitiveLog,
  ServiceQuotaExceededException: () => ServiceQuotaExceededException,
  Source: () => Source,
  ThrottlingException: () => ThrottlingException,
  Trace: () => Trace,
  TraceFilterSensitiveLog: () => TraceFilterSensitiveLog,
  TracePartFilterSensitiveLog: () => TracePartFilterSensitiveLog,
  Type: () => Type,
  ValidationException: () => ValidationException,
  __Client: () => import_smithy_client.Client,
  paginateRetrieve: () => paginateRetrieve
});
module.exports = __toCommonJS(src_exports);

// src/BedrockAgentRuntimeClient.ts
var import_middleware_host_header = require("@aws-sdk/middleware-host-header");
var import_middleware_logger = require("@aws-sdk/middleware-logger");
var import_middleware_recursion_detection = require("@aws-sdk/middleware-recursion-detection");
var import_middleware_user_agent = require("@aws-sdk/middleware-user-agent");
var import_config_resolver = require("@smithy/config-resolver");
var import_core = require("@smithy/core");
var import_eventstream_serde_config_resolver = require("@smithy/eventstream-serde-config-resolver");
var import_middleware_content_length = require("@smithy/middleware-content-length");
var import_middleware_endpoint = require("@smithy/middleware-endpoint");
var import_middleware_retry = require("@smithy/middleware-retry");

var import_httpAuthSchemeProvider = require("./auth/httpAuthSchemeProvider");

// src/endpoint/EndpointParameters.ts
var resolveClientEndpointParameters = /* @__PURE__ */ __name((options) => {
  return {
    ...options,
    useDualstackEndpoint: options.useDualstackEndpoint ?? false,
    useFipsEndpoint: options.useFipsEndpoint ?? false,
    defaultSigningName: "bedrock"
  };
}, "resolveClientEndpointParameters");
var commonParams = {
  UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
  Endpoint: { type: "builtInParams", name: "endpoint" },
  Region: { type: "builtInParams", name: "region" },
  UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" }
};

// src/BedrockAgentRuntimeClient.ts
var import_runtimeConfig = require("././runtimeConfig");

// src/runtimeExtensions.ts
var import_region_config_resolver = require("@aws-sdk/region-config-resolver");
var import_protocol_http = require("@smithy/protocol-http");
var import_smithy_client = require("@smithy/smithy-client");

// src/auth/httpAuthExtensionConfiguration.ts
var getHttpAuthExtensionConfiguration = /* @__PURE__ */ __name((runtimeConfig) => {
  const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
  let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
  let _credentials = runtimeConfig.credentials;
  return {
    setHttpAuthScheme(httpAuthScheme) {
      const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
      if (index === -1) {
        _httpAuthSchemes.push(httpAuthScheme);
      } else {
        _httpAuthSchemes.splice(index, 1, httpAuthScheme);
      }
    },
    httpAuthSchemes() {
      return _httpAuthSchemes;
    },
    setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
      _httpAuthSchemeProvider = httpAuthSchemeProvider;
    },
    httpAuthSchemeProvider() {
      return _httpAuthSchemeProvider;
    },
    setCredentials(credentials) {
      _credentials = credentials;
    },
    credentials() {
      return _credentials;
    }
  };
}, "getHttpAuthExtensionConfiguration");
var resolveHttpAuthRuntimeConfig = /* @__PURE__ */ __name((config) => {
  return {
    httpAuthSchemes: config.httpAuthSchemes(),
    httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
    credentials: config.credentials()
  };
}, "resolveHttpAuthRuntimeConfig");

// src/runtimeExtensions.ts
var asPartial = /* @__PURE__ */ __name((t) => t, "asPartial");
var resolveRuntimeExtensions = /* @__PURE__ */ __name((runtimeConfig, extensions) => {
  const extensionConfiguration = {
    ...asPartial((0, import_region_config_resolver.getAwsRegionExtensionConfiguration)(runtimeConfig)),
    ...asPartial((0, import_smithy_client.getDefaultExtensionConfiguration)(runtimeConfig)),
    ...asPartial((0, import_protocol_http.getHttpHandlerExtensionConfiguration)(runtimeConfig)),
    ...asPartial(getHttpAuthExtensionConfiguration(runtimeConfig))
  };
  extensions.forEach((extension) => extension.configure(extensionConfiguration));
  return {
    ...runtimeConfig,
    ...(0, import_region_config_resolver.resolveAwsRegionExtensionConfiguration)(extensionConfiguration),
    ...(0, import_smithy_client.resolveDefaultRuntimeConfig)(extensionConfiguration),
    ...(0, import_protocol_http.resolveHttpHandlerRuntimeConfig)(extensionConfiguration),
    ...resolveHttpAuthRuntimeConfig(extensionConfiguration)
  };
}, "resolveRuntimeExtensions");

// src/BedrockAgentRuntimeClient.ts
var _BedrockAgentRuntimeClient = class _BedrockAgentRuntimeClient extends import_smithy_client.Client {
  constructor(...[configuration]) {
    const _config_0 = (0, import_runtimeConfig.getRuntimeConfig)(configuration || {});
    const _config_1 = resolveClientEndpointParameters(_config_0);
    const _config_2 = (0, import_config_resolver.resolveRegionConfig)(_config_1);
    const _config_3 = (0, import_middleware_endpoint.resolveEndpointConfig)(_config_2);
    const _config_4 = (0, import_middleware_retry.resolveRetryConfig)(_config_3);
    const _config_5 = (0, import_middleware_host_header.resolveHostHeaderConfig)(_config_4);
    const _config_6 = (0, import_middleware_user_agent.resolveUserAgentConfig)(_config_5);
    const _config_7 = (0, import_eventstream_serde_config_resolver.resolveEventStreamSerdeConfig)(_config_6);
    const _config_8 = (0, import_httpAuthSchemeProvider.resolveHttpAuthSchemeConfig)(_config_7);
    const _config_9 = resolveRuntimeExtensions(_config_8, (configuration == null ? void 0 : configuration.extensions) || []);
    super(_config_9);
    this.config = _config_9;
    this.middlewareStack.use((0, import_middleware_retry.getRetryPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_content_length.getContentLengthPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_host_header.getHostHeaderPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_logger.getLoggerPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_recursion_detection.getRecursionDetectionPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_user_agent.getUserAgentPlugin)(this.config));
    this.middlewareStack.use(
      (0, import_core.getHttpAuthSchemeEndpointRuleSetPlugin)(this.config, {
        httpAuthSchemeParametersProvider: this.getDefaultHttpAuthSchemeParametersProvider(),
        identityProviderConfigProvider: this.getIdentityProviderConfigProvider()
      })
    );
    this.middlewareStack.use((0, import_core.getHttpSigningPlugin)(this.config));
  }
  /**
   * Destroy underlying resources, like sockets. It's usually not necessary to do this.
   * However in Node.js, it's best to explicitly shut down the client's agent when it is no longer needed.
   * Otherwise, sockets might stay open for quite a long time before the server terminates them.
   */
  destroy() {
    super.destroy();
  }
  getDefaultHttpAuthSchemeParametersProvider() {
    return import_httpAuthSchemeProvider.defaultBedrockAgentRuntimeHttpAuthSchemeParametersProvider;
  }
  getIdentityProviderConfigProvider() {
    return async (config) => new import_core.DefaultIdentityProviderConfig({
      "aws.auth#sigv4": config.credentials
    });
  }
};
__name(_BedrockAgentRuntimeClient, "BedrockAgentRuntimeClient");
var BedrockAgentRuntimeClient = _BedrockAgentRuntimeClient;

// src/BedrockAgentRuntime.ts


// src/commands/InvokeAgentCommand.ts

var import_middleware_serde = require("@smithy/middleware-serde");

var import_types = require("@smithy/types");

// src/models/models_0.ts


// src/models/BedrockAgentRuntimeServiceException.ts

var _BedrockAgentRuntimeServiceException = class _BedrockAgentRuntimeServiceException extends import_smithy_client.ServiceException {
  /**
   * @internal
   */
  constructor(options) {
    super(options);
    Object.setPrototypeOf(this, _BedrockAgentRuntimeServiceException.prototype);
  }
};
__name(_BedrockAgentRuntimeServiceException, "BedrockAgentRuntimeServiceException");
var BedrockAgentRuntimeServiceException = _BedrockAgentRuntimeServiceException;

// src/models/models_0.ts
var _AccessDeniedException = class _AccessDeniedException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "AccessDeniedException",
      $fault: "client",
      ...opts
    });
    this.name = "AccessDeniedException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _AccessDeniedException.prototype);
  }
};
__name(_AccessDeniedException, "AccessDeniedException");
var AccessDeniedException = _AccessDeniedException;
var _BadGatewayException = class _BadGatewayException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "BadGatewayException",
      $fault: "server",
      ...opts
    });
    this.name = "BadGatewayException";
    this.$fault = "server";
    Object.setPrototypeOf(this, _BadGatewayException.prototype);
    this.resourceName = opts.resourceName;
  }
};
__name(_BadGatewayException, "BadGatewayException");
var BadGatewayException = _BadGatewayException;
var _ConflictException = class _ConflictException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ConflictException",
      $fault: "client",
      ...opts
    });
    this.name = "ConflictException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ConflictException.prototype);
  }
};
__name(_ConflictException, "ConflictException");
var ConflictException = _ConflictException;
var _DependencyFailedException = class _DependencyFailedException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "DependencyFailedException",
      $fault: "client",
      ...opts
    });
    this.name = "DependencyFailedException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _DependencyFailedException.prototype);
    this.resourceName = opts.resourceName;
  }
};
__name(_DependencyFailedException, "DependencyFailedException");
var DependencyFailedException = _DependencyFailedException;
var _InternalServerException = class _InternalServerException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "InternalServerException",
      $fault: "server",
      ...opts
    });
    this.name = "InternalServerException";
    this.$fault = "server";
    Object.setPrototypeOf(this, _InternalServerException.prototype);
  }
};
__name(_InternalServerException, "InternalServerException");
var InternalServerException = _InternalServerException;
var RetrievalResultLocationType = {
  S3: "S3"
};
var _ResourceNotFoundException = class _ResourceNotFoundException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ResourceNotFoundException",
      $fault: "client",
      ...opts
    });
    this.name = "ResourceNotFoundException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ResourceNotFoundException.prototype);
  }
};
__name(_ResourceNotFoundException, "ResourceNotFoundException");
var ResourceNotFoundException = _ResourceNotFoundException;
var _ServiceQuotaExceededException = class _ServiceQuotaExceededException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ServiceQuotaExceededException",
      $fault: "client",
      ...opts
    });
    this.name = "ServiceQuotaExceededException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ServiceQuotaExceededException.prototype);
  }
};
__name(_ServiceQuotaExceededException, "ServiceQuotaExceededException");
var ServiceQuotaExceededException = _ServiceQuotaExceededException;
var _ThrottlingException = class _ThrottlingException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ThrottlingException",
      $fault: "client",
      ...opts
    });
    this.name = "ThrottlingException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ThrottlingException.prototype);
  }
};
__name(_ThrottlingException, "ThrottlingException");
var ThrottlingException = _ThrottlingException;
var InvocationType = {
  ACTION_GROUP: "ACTION_GROUP",
  FINISH: "FINISH",
  KNOWLEDGE_BASE: "KNOWLEDGE_BASE"
};
var CreationMode = {
  DEFAULT: "DEFAULT",
  OVERRIDDEN: "OVERRIDDEN"
};
var PromptType = {
  KNOWLEDGE_BASE_RESPONSE_GENERATION: "KNOWLEDGE_BASE_RESPONSE_GENERATION",
  ORCHESTRATION: "ORCHESTRATION",
  POST_PROCESSING: "POST_PROCESSING",
  PRE_PROCESSING: "PRE_PROCESSING"
};
var Source = {
  ACTION_GROUP: "ACTION_GROUP",
  KNOWLEDGE_BASE: "KNOWLEDGE_BASE",
  PARSER: "PARSER"
};
var Type = {
  ACTION_GROUP: "ACTION_GROUP",
  ASK_USER: "ASK_USER",
  FINISH: "FINISH",
  KNOWLEDGE_BASE: "KNOWLEDGE_BASE",
  REPROMPT: "REPROMPT"
};
var OrchestrationTrace;
((OrchestrationTrace3) => {
  OrchestrationTrace3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.rationale !== void 0)
      return visitor.rationale(value.rationale);
    if (value.invocationInput !== void 0)
      return visitor.invocationInput(value.invocationInput);
    if (value.observation !== void 0)
      return visitor.observation(value.observation);
    if (value.modelInvocationInput !== void 0)
      return visitor.modelInvocationInput(value.modelInvocationInput);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(OrchestrationTrace || (OrchestrationTrace = {}));
var PostProcessingTrace;
((PostProcessingTrace3) => {
  PostProcessingTrace3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.modelInvocationInput !== void 0)
      return visitor.modelInvocationInput(value.modelInvocationInput);
    if (value.modelInvocationOutput !== void 0)
      return visitor.modelInvocationOutput(value.modelInvocationOutput);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(PostProcessingTrace || (PostProcessingTrace = {}));
var PreProcessingTrace;
((PreProcessingTrace3) => {
  PreProcessingTrace3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.modelInvocationInput !== void 0)
      return visitor.modelInvocationInput(value.modelInvocationInput);
    if (value.modelInvocationOutput !== void 0)
      return visitor.modelInvocationOutput(value.modelInvocationOutput);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(PreProcessingTrace || (PreProcessingTrace = {}));
var Trace;
((Trace3) => {
  Trace3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.preProcessingTrace !== void 0)
      return visitor.preProcessingTrace(value.preProcessingTrace);
    if (value.orchestrationTrace !== void 0)
      return visitor.orchestrationTrace(value.orchestrationTrace);
    if (value.postProcessingTrace !== void 0)
      return visitor.postProcessingTrace(value.postProcessingTrace);
    if (value.failureTrace !== void 0)
      return visitor.failureTrace(value.failureTrace);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(Trace || (Trace = {}));
var _ValidationException = class _ValidationException extends BedrockAgentRuntimeServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ValidationException",
      $fault: "client",
      ...opts
    });
    this.name = "ValidationException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ValidationException.prototype);
  }
};
__name(_ValidationException, "ValidationException");
var ValidationException = _ValidationException;
var ResponseStream;
((ResponseStream3) => {
  ResponseStream3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.chunk !== void 0)
      return visitor.chunk(value.chunk);
    if (value.trace !== void 0)
      return visitor.trace(value.trace);
    if (value.internalServerException !== void 0)
      return visitor.internalServerException(value.internalServerException);
    if (value.validationException !== void 0)
      return visitor.validationException(value.validationException);
    if (value.resourceNotFoundException !== void 0)
      return visitor.resourceNotFoundException(value.resourceNotFoundException);
    if (value.serviceQuotaExceededException !== void 0)
      return visitor.serviceQuotaExceededException(value.serviceQuotaExceededException);
    if (value.throttlingException !== void 0)
      return visitor.throttlingException(value.throttlingException);
    if (value.accessDeniedException !== void 0)
      return visitor.accessDeniedException(value.accessDeniedException);
    if (value.conflictException !== void 0)
      return visitor.conflictException(value.conflictException);
    if (value.dependencyFailedException !== void 0)
      return visitor.dependencyFailedException(value.dependencyFailedException);
    if (value.badGatewayException !== void 0)
      return visitor.badGatewayException(value.badGatewayException);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(ResponseStream || (ResponseStream = {}));
var RetrieveAndGenerateType = {
  KNOWLEDGE_BASE: "KNOWLEDGE_BASE"
};
var ActionGroupInvocationInputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupName && { actionGroupName: import_smithy_client.SENSITIVE_STRING },
  ...obj.verb && { verb: import_smithy_client.SENSITIVE_STRING },
  ...obj.apiPath && { apiPath: import_smithy_client.SENSITIVE_STRING }
}), "ActionGroupInvocationInputFilterSensitiveLog");
var ActionGroupInvocationOutputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING }
}), "ActionGroupInvocationOutputFilterSensitiveLog");
var InvokeAgentRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.inputText && { inputText: import_smithy_client.SENSITIVE_STRING }
}), "InvokeAgentRequestFilterSensitiveLog");
var PayloadPartFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.bytes && { bytes: import_smithy_client.SENSITIVE_STRING }
}), "PayloadPartFilterSensitiveLog");
var FailureTraceFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.failureReason && { failureReason: import_smithy_client.SENSITIVE_STRING }
}), "FailureTraceFilterSensitiveLog");
var KnowledgeBaseLookupInputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING },
  ...obj.knowledgeBaseId && { knowledgeBaseId: import_smithy_client.SENSITIVE_STRING }
}), "KnowledgeBaseLookupInputFilterSensitiveLog");
var InvocationInputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupInvocationInput && {
    actionGroupInvocationInput: ActionGroupInvocationInputFilterSensitiveLog(obj.actionGroupInvocationInput)
  },
  ...obj.knowledgeBaseLookupInput && {
    knowledgeBaseLookupInput: KnowledgeBaseLookupInputFilterSensitiveLog(obj.knowledgeBaseLookupInput)
  }
}), "InvocationInputFilterSensitiveLog");
var ModelInvocationInputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING }
}), "ModelInvocationInputFilterSensitiveLog");
var FinalResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING }
}), "FinalResponseFilterSensitiveLog");
var RepromptResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.source && { source: import_smithy_client.SENSITIVE_STRING }
}), "RepromptResponseFilterSensitiveLog");
var ObservationFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupInvocationOutput && {
    actionGroupInvocationOutput: ActionGroupInvocationOutputFilterSensitiveLog(obj.actionGroupInvocationOutput)
  },
  ...obj.finalResponse && { finalResponse: FinalResponseFilterSensitiveLog(obj.finalResponse) },
  ...obj.repromptResponse && { repromptResponse: import_smithy_client.SENSITIVE_STRING }
}), "ObservationFilterSensitiveLog");
var RationaleFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING }
}), "RationaleFilterSensitiveLog");
var OrchestrationTraceFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.rationale !== void 0)
    return { rationale: import_smithy_client.SENSITIVE_STRING };
  if (obj.invocationInput !== void 0)
    return { invocationInput: import_smithy_client.SENSITIVE_STRING };
  if (obj.observation !== void 0)
    return { observation: import_smithy_client.SENSITIVE_STRING };
  if (obj.modelInvocationInput !== void 0)
    return { modelInvocationInput: import_smithy_client.SENSITIVE_STRING };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "OrchestrationTraceFilterSensitiveLog");
var PostProcessingParsedResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.text && { text: import_smithy_client.SENSITIVE_STRING }
}), "PostProcessingParsedResponseFilterSensitiveLog");
var PostProcessingModelInvocationOutputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.parsedResponse && { parsedResponse: import_smithy_client.SENSITIVE_STRING }
}), "PostProcessingModelInvocationOutputFilterSensitiveLog");
var PostProcessingTraceFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.modelInvocationInput !== void 0)
    return { modelInvocationInput: import_smithy_client.SENSITIVE_STRING };
  if (obj.modelInvocationOutput !== void 0)
    return { modelInvocationOutput: import_smithy_client.SENSITIVE_STRING };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "PostProcessingTraceFilterSensitiveLog");
var PreProcessingParsedResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.rationale && { rationale: import_smithy_client.SENSITIVE_STRING }
}), "PreProcessingParsedResponseFilterSensitiveLog");
var PreProcessingModelInvocationOutputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.parsedResponse && { parsedResponse: import_smithy_client.SENSITIVE_STRING }
}), "PreProcessingModelInvocationOutputFilterSensitiveLog");
var PreProcessingTraceFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.modelInvocationInput !== void 0)
    return { modelInvocationInput: import_smithy_client.SENSITIVE_STRING };
  if (obj.modelInvocationOutput !== void 0)
    return { modelInvocationOutput: import_smithy_client.SENSITIVE_STRING };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "PreProcessingTraceFilterSensitiveLog");
var TraceFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.preProcessingTrace !== void 0)
    return { preProcessingTrace: import_smithy_client.SENSITIVE_STRING };
  if (obj.orchestrationTrace !== void 0)
    return { orchestrationTrace: import_smithy_client.SENSITIVE_STRING };
  if (obj.postProcessingTrace !== void 0)
    return { postProcessingTrace: import_smithy_client.SENSITIVE_STRING };
  if (obj.failureTrace !== void 0)
    return { failureTrace: import_smithy_client.SENSITIVE_STRING };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "TraceFilterSensitiveLog");
var TracePartFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.trace && { trace: import_smithy_client.SENSITIVE_STRING }
}), "TracePartFilterSensitiveLog");
var ResponseStreamFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.chunk !== void 0)
    return { chunk: import_smithy_client.SENSITIVE_STRING };
  if (obj.trace !== void 0)
    return { trace: import_smithy_client.SENSITIVE_STRING };
  if (obj.internalServerException !== void 0)
    return { internalServerException: obj.internalServerException };
  if (obj.validationException !== void 0)
    return { validationException: obj.validationException };
  if (obj.resourceNotFoundException !== void 0)
    return { resourceNotFoundException: obj.resourceNotFoundException };
  if (obj.serviceQuotaExceededException !== void 0)
    return { serviceQuotaExceededException: obj.serviceQuotaExceededException };
  if (obj.throttlingException !== void 0)
    return { throttlingException: obj.throttlingException };
  if (obj.accessDeniedException !== void 0)
    return { accessDeniedException: obj.accessDeniedException };
  if (obj.conflictException !== void 0)
    return { conflictException: obj.conflictException };
  if (obj.dependencyFailedException !== void 0)
    return { dependencyFailedException: obj.dependencyFailedException };
  if (obj.badGatewayException !== void 0)
    return { badGatewayException: obj.badGatewayException };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "ResponseStreamFilterSensitiveLog");
var InvokeAgentResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.completion && { completion: "STREAMING_CONTENT" }
}), "InvokeAgentResponseFilterSensitiveLog");
var RetrieveAndGenerateInputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj
}), "RetrieveAndGenerateInputFilterSensitiveLog");
var RetrieveAndGenerateRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.input && { input: import_smithy_client.SENSITIVE_STRING }
}), "RetrieveAndGenerateRequestFilterSensitiveLog");
var RetrieveAndGenerateOutputFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj
}), "RetrieveAndGenerateOutputFilterSensitiveLog");
var RetrieveAndGenerateResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.output && { output: import_smithy_client.SENSITIVE_STRING }
}), "RetrieveAndGenerateResponseFilterSensitiveLog");
var KnowledgeBaseQueryFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj
}), "KnowledgeBaseQueryFilterSensitiveLog");
var RetrieveRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.retrievalQuery && { retrievalQuery: import_smithy_client.SENSITIVE_STRING }
}), "RetrieveRequestFilterSensitiveLog");
var RetrieveResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.retrievalResults && { retrievalResults: import_smithy_client.SENSITIVE_STRING }
}), "RetrieveResponseFilterSensitiveLog");

// src/protocols/Aws_restJson1.ts
var import_core2 = require("@aws-sdk/core");


var se_InvokeAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentAliases/{agentAliasId}/sessions/{sessionId}/text");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentAliasId", () => input.agentAliasId, "{agentAliasId}", false);
  b.p("sessionId", () => input.sessionId, "{sessionId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      enableTrace: [],
      endSession: [],
      inputText: [],
      sessionState: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_InvokeAgentCommand");
var se_RetrieveCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/retrieve");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      nextToken: [],
      retrievalConfiguration: (_) => (0, import_smithy_client._json)(_),
      retrievalQuery: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_RetrieveCommand");
var se_RetrieveAndGenerateCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/retrieveAndGenerate");
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      input: (_) => (0, import_smithy_client._json)(_),
      retrieveAndGenerateConfiguration: (_) => (0, import_smithy_client._json)(_),
      sessionConfiguration: (_) => (0, import_smithy_client._json)(_),
      sessionId: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_RetrieveAndGenerateCommand");
var de_InvokeAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output),
    [_cT]: [, output.headers[_xabact]],
    [_sI]: [, output.headers[_xabasi]]
  });
  const data = output.body;
  contents.completion = de_ResponseStream(data, context);
  return contents;
}, "de_InvokeAgentCommand");
var de_RetrieveCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    nextToken: import_smithy_client.expectString,
    retrievalResults: (_) => de_KnowledgeBaseRetrievalResults(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_RetrieveCommand");
var de_RetrieveAndGenerateCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    citations: import_smithy_client._json,
    output: import_smithy_client._json,
    sessionId: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_RetrieveAndGenerateCommand");
var de_CommandError = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseErrorBody(output.body, context)
  };
  const errorCode = loadRestJsonErrorCode(output, parsedOutput.body);
  switch (errorCode) {
    case "AccessDeniedException":
    case "com.amazonaws.bedrockagentruntime#AccessDeniedException":
      throw await de_AccessDeniedExceptionRes(parsedOutput, context);
    case "BadGatewayException":
    case "com.amazonaws.bedrockagentruntime#BadGatewayException":
      throw await de_BadGatewayExceptionRes(parsedOutput, context);
    case "ConflictException":
    case "com.amazonaws.bedrockagentruntime#ConflictException":
      throw await de_ConflictExceptionRes(parsedOutput, context);
    case "DependencyFailedException":
    case "com.amazonaws.bedrockagentruntime#DependencyFailedException":
      throw await de_DependencyFailedExceptionRes(parsedOutput, context);
    case "InternalServerException":
    case "com.amazonaws.bedrockagentruntime#InternalServerException":
      throw await de_InternalServerExceptionRes(parsedOutput, context);
    case "ResourceNotFoundException":
    case "com.amazonaws.bedrockagentruntime#ResourceNotFoundException":
      throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
    case "ServiceQuotaExceededException":
    case "com.amazonaws.bedrockagentruntime#ServiceQuotaExceededException":
      throw await de_ServiceQuotaExceededExceptionRes(parsedOutput, context);
    case "ThrottlingException":
    case "com.amazonaws.bedrockagentruntime#ThrottlingException":
      throw await de_ThrottlingExceptionRes(parsedOutput, context);
    case "ValidationException":
    case "com.amazonaws.bedrockagentruntime#ValidationException":
      throw await de_ValidationExceptionRes(parsedOutput, context);
    default:
      const parsedBody = parsedOutput.body;
      return throwDefaultError({
        output,
        parsedBody,
        errorCode
      });
  }
}, "de_CommandError");
var throwDefaultError = (0, import_smithy_client.withBaseException)(BedrockAgentRuntimeServiceException);
var de_AccessDeniedExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new AccessDeniedException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_AccessDeniedExceptionRes");
var de_BadGatewayExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString,
    resourceName: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new BadGatewayException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_BadGatewayExceptionRes");
var de_ConflictExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ConflictException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ConflictExceptionRes");
var de_DependencyFailedExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString,
    resourceName: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new DependencyFailedException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_DependencyFailedExceptionRes");
var de_InternalServerExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new InternalServerException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_InternalServerExceptionRes");
var de_ResourceNotFoundExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ResourceNotFoundException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ResourceNotFoundExceptionRes");
var de_ServiceQuotaExceededExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ServiceQuotaExceededException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ServiceQuotaExceededExceptionRes");
var de_ThrottlingExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ThrottlingException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ThrottlingExceptionRes");
var de_ValidationExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ValidationException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ValidationExceptionRes");
var de_ResponseStream = /* @__PURE__ */ __name((output, context) => {
  return context.eventStreamMarshaller.deserialize(output, async (event) => {
    if (event["chunk"] != null) {
      return {
        chunk: await de_PayloadPart_event(event["chunk"], context)
      };
    }
    if (event["trace"] != null) {
      return {
        trace: await de_TracePart_event(event["trace"], context)
      };
    }
    if (event["internalServerException"] != null) {
      return {
        internalServerException: await de_InternalServerException_event(event["internalServerException"], context)
      };
    }
    if (event["validationException"] != null) {
      return {
        validationException: await de_ValidationException_event(event["validationException"], context)
      };
    }
    if (event["resourceNotFoundException"] != null) {
      return {
        resourceNotFoundException: await de_ResourceNotFoundException_event(
          event["resourceNotFoundException"],
          context
        )
      };
    }
    if (event["serviceQuotaExceededException"] != null) {
      return {
        serviceQuotaExceededException: await de_ServiceQuotaExceededException_event(
          event["serviceQuotaExceededException"],
          context
        )
      };
    }
    if (event["throttlingException"] != null) {
      return {
        throttlingException: await de_ThrottlingException_event(event["throttlingException"], context)
      };
    }
    if (event["accessDeniedException"] != null) {
      return {
        accessDeniedException: await de_AccessDeniedException_event(event["accessDeniedException"], context)
      };
    }
    if (event["conflictException"] != null) {
      return {
        conflictException: await de_ConflictException_event(event["conflictException"], context)
      };
    }
    if (event["dependencyFailedException"] != null) {
      return {
        dependencyFailedException: await de_DependencyFailedException_event(
          event["dependencyFailedException"],
          context
        )
      };
    }
    if (event["badGatewayException"] != null) {
      return {
        badGatewayException: await de_BadGatewayException_event(event["badGatewayException"], context)
      };
    }
    return { $unknown: output };
  });
}, "de_ResponseStream");
var de_AccessDeniedException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_AccessDeniedExceptionRes(parsedOutput, context);
}, "de_AccessDeniedException_event");
var de_BadGatewayException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_BadGatewayExceptionRes(parsedOutput, context);
}, "de_BadGatewayException_event");
var de_ConflictException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_ConflictExceptionRes(parsedOutput, context);
}, "de_ConflictException_event");
var de_DependencyFailedException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_DependencyFailedExceptionRes(parsedOutput, context);
}, "de_DependencyFailedException_event");
var de_InternalServerException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_InternalServerExceptionRes(parsedOutput, context);
}, "de_InternalServerException_event");
var de_PayloadPart_event = /* @__PURE__ */ __name(async (output, context) => {
  const contents = {};
  const data = await parseBody(output.body, context);
  Object.assign(contents, de_PayloadPart(data, context));
  return contents;
}, "de_PayloadPart_event");
var de_ResourceNotFoundException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_ResourceNotFoundExceptionRes(parsedOutput, context);
}, "de_ResourceNotFoundException_event");
var de_ServiceQuotaExceededException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_ServiceQuotaExceededExceptionRes(parsedOutput, context);
}, "de_ServiceQuotaExceededException_event");
var de_ThrottlingException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_ThrottlingExceptionRes(parsedOutput, context);
}, "de_ThrottlingException_event");
var de_TracePart_event = /* @__PURE__ */ __name(async (output, context) => {
  const contents = {};
  const data = await parseBody(output.body, context);
  Object.assign(contents, de_TracePart(data, context));
  return contents;
}, "de_TracePart_event");
var de_ValidationException_event = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseBody(output.body, context)
  };
  return de_ValidationExceptionRes(parsedOutput, context);
}, "de_ValidationException_event");
var de_InferenceConfiguration = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    maximumLength: import_smithy_client.expectInt32,
    stopSequences: import_smithy_client._json,
    temperature: import_smithy_client.limitedParseFloat32,
    topK: import_smithy_client.expectInt32,
    topP: import_smithy_client.limitedParseFloat32
  });
}, "de_InferenceConfiguration");
var de_KnowledgeBaseRetrievalResult = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    content: import_smithy_client._json,
    location: import_smithy_client._json,
    score: import_smithy_client.limitedParseDouble
  });
}, "de_KnowledgeBaseRetrievalResult");
var de_KnowledgeBaseRetrievalResults = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_KnowledgeBaseRetrievalResult(entry, context);
  });
  return retVal;
}, "de_KnowledgeBaseRetrievalResults");
var de_ModelInvocationInput = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    inferenceConfiguration: (_) => de_InferenceConfiguration(_, context),
    overrideLambda: import_smithy_client.expectString,
    parserMode: import_smithy_client.expectString,
    promptCreationMode: import_smithy_client.expectString,
    text: import_smithy_client.expectString,
    traceId: import_smithy_client.expectString,
    type: import_smithy_client.expectString
  });
}, "de_ModelInvocationInput");
var de_OrchestrationTrace = /* @__PURE__ */ __name((output, context) => {
  if (output.invocationInput != null) {
    return {
      invocationInput: (0, import_smithy_client._json)(output.invocationInput)
    };
  }
  if (output.modelInvocationInput != null) {
    return {
      modelInvocationInput: de_ModelInvocationInput(output.modelInvocationInput, context)
    };
  }
  if (output.observation != null) {
    return {
      observation: (0, import_smithy_client._json)(output.observation)
    };
  }
  if (output.rationale != null) {
    return {
      rationale: (0, import_smithy_client._json)(output.rationale)
    };
  }
  return { $unknown: Object.entries(output)[0] };
}, "de_OrchestrationTrace");
var de_PayloadPart = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    attribution: import_smithy_client._json,
    bytes: context.base64Decoder
  });
}, "de_PayloadPart");
var de_PostProcessingTrace = /* @__PURE__ */ __name((output, context) => {
  if (output.modelInvocationInput != null) {
    return {
      modelInvocationInput: de_ModelInvocationInput(output.modelInvocationInput, context)
    };
  }
  if (output.modelInvocationOutput != null) {
    return {
      modelInvocationOutput: (0, import_smithy_client._json)(output.modelInvocationOutput)
    };
  }
  return { $unknown: Object.entries(output)[0] };
}, "de_PostProcessingTrace");
var de_PreProcessingTrace = /* @__PURE__ */ __name((output, context) => {
  if (output.modelInvocationInput != null) {
    return {
      modelInvocationInput: de_ModelInvocationInput(output.modelInvocationInput, context)
    };
  }
  if (output.modelInvocationOutput != null) {
    return {
      modelInvocationOutput: (0, import_smithy_client._json)(output.modelInvocationOutput)
    };
  }
  return { $unknown: Object.entries(output)[0] };
}, "de_PreProcessingTrace");
var de_Trace = /* @__PURE__ */ __name((output, context) => {
  if (output.failureTrace != null) {
    return {
      failureTrace: (0, import_smithy_client._json)(output.failureTrace)
    };
  }
  if (output.orchestrationTrace != null) {
    return {
      orchestrationTrace: de_OrchestrationTrace((0, import_core2.awsExpectUnion)(output.orchestrationTrace), context)
    };
  }
  if (output.postProcessingTrace != null) {
    return {
      postProcessingTrace: de_PostProcessingTrace((0, import_core2.awsExpectUnion)(output.postProcessingTrace), context)
    };
  }
  if (output.preProcessingTrace != null) {
    return {
      preProcessingTrace: de_PreProcessingTrace((0, import_core2.awsExpectUnion)(output.preProcessingTrace), context)
    };
  }
  return { $unknown: Object.entries(output)[0] };
}, "de_Trace");
var de_TracePart = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentAliasId: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString,
    sessionId: import_smithy_client.expectString,
    trace: (_) => de_Trace((0, import_core2.awsExpectUnion)(_), context)
  });
}, "de_TracePart");
var deserializeMetadata = /* @__PURE__ */ __name((output) => ({
  httpStatusCode: output.statusCode,
  requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
  extendedRequestId: output.headers["x-amz-id-2"],
  cfId: output.headers["x-amz-cf-id"]
}), "deserializeMetadata");
var collectBodyString = /* @__PURE__ */ __name((streamBody, context) => (0, import_smithy_client.collectBody)(streamBody, context).then((body) => context.utf8Encoder(body)), "collectBodyString");
var _cT = "contentType";
var _sI = "sessionId";
var _xabact = "x-amzn-bedrock-agent-content-type";
var _xabasi = "x-amz-bedrock-agent-session-id";
var parseBody = /* @__PURE__ */ __name((streamBody, context) => collectBodyString(streamBody, context).then((encoded) => {
  if (encoded.length) {
    return JSON.parse(encoded);
  }
  return {};
}), "parseBody");
var parseErrorBody = /* @__PURE__ */ __name(async (errorBody, context) => {
  const value = await parseBody(errorBody, context);
  value.message = value.message ?? value.Message;
  return value;
}, "parseErrorBody");
var loadRestJsonErrorCode = /* @__PURE__ */ __name((output, data) => {
  const findKey = /* @__PURE__ */ __name((object, key) => Object.keys(object).find((k) => k.toLowerCase() === key.toLowerCase()), "findKey");
  const sanitizeErrorCode = /* @__PURE__ */ __name((rawValue) => {
    let cleanValue = rawValue;
    if (typeof cleanValue === "number") {
      cleanValue = cleanValue.toString();
    }
    if (cleanValue.indexOf(",") >= 0) {
      cleanValue = cleanValue.split(",")[0];
    }
    if (cleanValue.indexOf(":") >= 0) {
      cleanValue = cleanValue.split(":")[0];
    }
    if (cleanValue.indexOf("#") >= 0) {
      cleanValue = cleanValue.split("#")[1];
    }
    return cleanValue;
  }, "sanitizeErrorCode");
  const headerKey = findKey(output.headers, "x-amzn-errortype");
  if (headerKey !== void 0) {
    return sanitizeErrorCode(output.headers[headerKey]);
  }
  if (data.code !== void 0) {
    return sanitizeErrorCode(data.code);
  }
  if (data["__type"] !== void 0) {
    return sanitizeErrorCode(data["__type"]);
  }
}, "loadRestJsonErrorCode");

// src/commands/InvokeAgentCommand.ts
var _InvokeAgentCommand = class _InvokeAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentRunTimeService", "InvokeAgent", {
  /**
   * @internal
   */
  eventStream: {
    output: true
  }
}).n("BedrockAgentRuntimeClient", "InvokeAgentCommand").f(InvokeAgentRequestFilterSensitiveLog, InvokeAgentResponseFilterSensitiveLog).ser(se_InvokeAgentCommand).de(de_InvokeAgentCommand).build() {
};
__name(_InvokeAgentCommand, "InvokeAgentCommand");
var InvokeAgentCommand = _InvokeAgentCommand;

// src/commands/RetrieveAndGenerateCommand.ts




var _RetrieveAndGenerateCommand = class _RetrieveAndGenerateCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentRunTimeService", "RetrieveAndGenerate", {}).n("BedrockAgentRuntimeClient", "RetrieveAndGenerateCommand").f(RetrieveAndGenerateRequestFilterSensitiveLog, RetrieveAndGenerateResponseFilterSensitiveLog).ser(se_RetrieveAndGenerateCommand).de(de_RetrieveAndGenerateCommand).build() {
};
__name(_RetrieveAndGenerateCommand, "RetrieveAndGenerateCommand");
var RetrieveAndGenerateCommand = _RetrieveAndGenerateCommand;

// src/commands/RetrieveCommand.ts




var _RetrieveCommand = class _RetrieveCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentRunTimeService", "Retrieve", {}).n("BedrockAgentRuntimeClient", "RetrieveCommand").f(RetrieveRequestFilterSensitiveLog, RetrieveResponseFilterSensitiveLog).ser(se_RetrieveCommand).de(de_RetrieveCommand).build() {
};
__name(_RetrieveCommand, "RetrieveCommand");
var RetrieveCommand = _RetrieveCommand;

// src/BedrockAgentRuntime.ts
var commands = {
  InvokeAgentCommand,
  RetrieveCommand,
  RetrieveAndGenerateCommand
};
var _BedrockAgentRuntime = class _BedrockAgentRuntime extends BedrockAgentRuntimeClient {
};
__name(_BedrockAgentRuntime, "BedrockAgentRuntime");
var BedrockAgentRuntime = _BedrockAgentRuntime;
(0, import_smithy_client.createAggregatedClient)(commands, BedrockAgentRuntime);

// src/pagination/RetrievePaginator.ts

var paginateRetrieve = (0, import_core.createPaginator)(BedrockAgentRuntimeClient, RetrieveCommand, "nextToken", "nextToken", "");

// src/index.ts
var import_util_endpoints = require("@aws-sdk/util-endpoints");
// Annotate the CommonJS export names for ESM import in node:

0 && (module.exports = {
  BedrockAgentRuntimeServiceException,
  __Client,
  BedrockAgentRuntimeClient,
  BedrockAgentRuntime,
  $Command,
  InvokeAgentCommand,
  RetrieveAndGenerateCommand,
  RetrieveCommand,
  paginateRetrieve,
  AccessDeniedException,
  BadGatewayException,
  ConflictException,
  DependencyFailedException,
  InternalServerException,
  RetrievalResultLocationType,
  ResourceNotFoundException,
  ServiceQuotaExceededException,
  ThrottlingException,
  InvocationType,
  CreationMode,
  PromptType,
  Source,
  Type,
  OrchestrationTrace,
  PostProcessingTrace,
  PreProcessingTrace,
  Trace,
  ValidationException,
  ResponseStream,
  RetrieveAndGenerateType,
  ActionGroupInvocationInputFilterSensitiveLog,
  ActionGroupInvocationOutputFilterSensitiveLog,
  InvokeAgentRequestFilterSensitiveLog,
  PayloadPartFilterSensitiveLog,
  FailureTraceFilterSensitiveLog,
  KnowledgeBaseLookupInputFilterSensitiveLog,
  InvocationInputFilterSensitiveLog,
  ModelInvocationInputFilterSensitiveLog,
  FinalResponseFilterSensitiveLog,
  RepromptResponseFilterSensitiveLog,
  ObservationFilterSensitiveLog,
  RationaleFilterSensitiveLog,
  OrchestrationTraceFilterSensitiveLog,
  PostProcessingParsedResponseFilterSensitiveLog,
  PostProcessingModelInvocationOutputFilterSensitiveLog,
  PostProcessingTraceFilterSensitiveLog,
  PreProcessingParsedResponseFilterSensitiveLog,
  PreProcessingModelInvocationOutputFilterSensitiveLog,
  PreProcessingTraceFilterSensitiveLog,
  TraceFilterSensitiveLog,
  TracePartFilterSensitiveLog,
  ResponseStreamFilterSensitiveLog,
  InvokeAgentResponseFilterSensitiveLog,
  RetrieveAndGenerateInputFilterSensitiveLog,
  RetrieveAndGenerateRequestFilterSensitiveLog,
  RetrieveAndGenerateOutputFilterSensitiveLog,
  RetrieveAndGenerateResponseFilterSensitiveLog,
  KnowledgeBaseQueryFilterSensitiveLog,
  RetrieveRequestFilterSensitiveLog,
  RetrieveResponseFilterSensitiveLog
});
