// @aws-amplify/notifications/dist/esm/inAppMessaging/providers/pinpoint/apis/syncMessages.mjs
import { InAppMessagingAction } from '@aws-amplify/core/internals/utils';
import { resolveEndpointId } from '@aws-amplify/core/internals/providers/pinpoint';
import { defaultStorage } from '@aws-amplify/core';
import { resolveConfig } from '../utils/resolveConfig.mjs';
import { resolveCredentials } from '../utils/resolveCredentials.mjs';
import { getInAppMessagingUserAgentString } from '../utils/userAgent.mjs';
import { PINPOINT_KEY_PREFIX, STORAGE_KEY_SUFFIX, CATEGORY, CHANNEL_TYPE } from '../utils/constants.mjs';
import '../utils/messageProcessingHelpers.mjs';
import { getInAppMessages } from '@aws-amplify/core/internals/aws-clients/pinpoint';
import '../../../errors/validation.mjs';
import { assertServiceError } from '../../../errors/assertServiceError.mjs';
import { assertIsInitialized } from '../../../utils/statusHelpers.mjs';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
/**
 * Fetch and persist messages from Pinpoint campaigns.
 * Calling this API is necessary to trigger InApp messages on the device.
 *
 * @throws service exceptions - Thrown when the underlying Pinpoint service returns an error.
 * @throws validation: {@link InAppMessagingValidationErrorCode} - Thrown when the provided parameters or library
 * configuration is incorrect, or if In App messaging hasn't been initialized.
 * @returns A promise that will resolve when the operation is complete.
 * @example
 * ```ts
 * // Sync InApp messages with Pinpoint and device.
 * await syncMessages();
 *
 * ```
 */
async function syncMessages() {
    assertIsInitialized();
    const messages = await fetchInAppMessages();
    if (!messages || messages.length === 0) {
        return;
    }
    try {
        const key = `${PINPOINT_KEY_PREFIX}${STORAGE_KEY_SUFFIX}`;
        await defaultStorage.setItem(key, JSON.stringify(messages));
    }
    catch (error) {
        assertServiceError(error);
        throw error;
    }
}
async function fetchInAppMessages() {
    try {
        const { credentials, identityId } = await resolveCredentials();
        const { appId, region } = resolveConfig();
        const endpointId = await resolveEndpointId({
            appId,
            category: CATEGORY,
            channelType: CHANNEL_TYPE,
            credentials,
            identityId,
            region,
            userAgentValue: getInAppMessagingUserAgentString(InAppMessagingAction.SyncMessages),
        });
        const input = {
            ApplicationId: appId,
            EndpointId: endpointId,
        };
        const response = await getInAppMessages({ credentials, region }, input);
        const { InAppMessageCampaigns: messages } = response.InAppMessagesResponse ?? {};
        return messages;
    }
    catch (error) {
        assertServiceError(error);
        throw error;
    }
}

export { syncMessages };
