// @aws-amplify/datastore/dist/esm/index.mjs
export { AsyncCollection, AsyncItem, DataStore, DataStoreClass, initSchema } from './datastore/datastore.mjs';
export { ModelPredicateCreator, Predicates } from './predicates/index.mjs';
import { USER, traverseModel, validatePredicate, isNonModelConstructor, isModelConstructor } from './util.mjs';
export { NAMESPACES } from './util.mjs';
export { AuthModeStrategyType, DISCARD, GraphQLScalarType, LimitTimerRaceResolvedValues, ModelAttributeAuthAllow, ModelAttributeAuthProvider, ModelOperation, OpType, PredicateInternalsKey, ProcessName, QueryOne, SortDirection, isAssociatedWith, isEnumFieldType, isFieldAssociation, isGraphQLScalarType, isIdentifierObject, isModelAttributeAuth, isModelAttributeCompositeKey, isModelAttributeKey, isModelAttributePrimaryKey, isModelFieldType, isNonModelFieldType, isPredicateGroup, isPredicateObj, isSchemaModel, isSchemaModelWithAttributes, isTargetNameAssociation, syncExpression } from './types.mjs';
export { ModelSortPredicateCreator } from './predicates/sort.mjs';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const utils = {
    USER,
    traverseModel,
    validatePredicate,
    isNonModelConstructor,
    isModelConstructor,
};

export { utils };
