// @aws-sdk/client-bedrock-agent/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  APISchema: () => APISchema,
  APISchemaFilterSensitiveLog: () => APISchemaFilterSensitiveLog,
  AccessDeniedException: () => AccessDeniedException,
  ActionGroupExecutor: () => ActionGroupExecutor,
  ActionGroupSignature: () => ActionGroupSignature,
  ActionGroupState: () => ActionGroupState,
  AgentActionGroupFilterSensitiveLog: () => AgentActionGroupFilterSensitiveLog,
  AgentAliasStatus: () => AgentAliasStatus,
  AgentFilterSensitiveLog: () => AgentFilterSensitiveLog,
  AgentStatus: () => AgentStatus,
  AgentVersionFilterSensitiveLog: () => AgentVersionFilterSensitiveLog,
  AssociateAgentKnowledgeBaseCommand: () => AssociateAgentKnowledgeBaseCommand,
  BedrockAgent: () => BedrockAgent,
  BedrockAgentClient: () => BedrockAgentClient,
  BedrockAgentServiceException: () => BedrockAgentServiceException,
  ChunkingStrategy: () => ChunkingStrategy,
  ConflictException: () => ConflictException,
  CreateAgentActionGroupCommand: () => CreateAgentActionGroupCommand,
  CreateAgentActionGroupRequestFilterSensitiveLog: () => CreateAgentActionGroupRequestFilterSensitiveLog,
  CreateAgentActionGroupResponseFilterSensitiveLog: () => CreateAgentActionGroupResponseFilterSensitiveLog,
  CreateAgentAliasCommand: () => CreateAgentAliasCommand,
  CreateAgentCommand: () => CreateAgentCommand,
  CreateAgentRequestFilterSensitiveLog: () => CreateAgentRequestFilterSensitiveLog,
  CreateAgentResponseFilterSensitiveLog: () => CreateAgentResponseFilterSensitiveLog,
  CreateDataSourceCommand: () => CreateDataSourceCommand,
  CreateKnowledgeBaseCommand: () => CreateKnowledgeBaseCommand,
  CreationMode: () => CreationMode,
  DataSourceStatus: () => DataSourceStatus,
  DataSourceType: () => DataSourceType,
  DeleteAgentActionGroupCommand: () => DeleteAgentActionGroupCommand,
  DeleteAgentAliasCommand: () => DeleteAgentAliasCommand,
  DeleteAgentCommand: () => DeleteAgentCommand,
  DeleteAgentVersionCommand: () => DeleteAgentVersionCommand,
  DeleteDataSourceCommand: () => DeleteDataSourceCommand,
  DeleteKnowledgeBaseCommand: () => DeleteKnowledgeBaseCommand,
  DisassociateAgentKnowledgeBaseCommand: () => DisassociateAgentKnowledgeBaseCommand,
  GetAgentActionGroupCommand: () => GetAgentActionGroupCommand,
  GetAgentActionGroupResponseFilterSensitiveLog: () => GetAgentActionGroupResponseFilterSensitiveLog,
  GetAgentAliasCommand: () => GetAgentAliasCommand,
  GetAgentCommand: () => GetAgentCommand,
  GetAgentKnowledgeBaseCommand: () => GetAgentKnowledgeBaseCommand,
  GetAgentResponseFilterSensitiveLog: () => GetAgentResponseFilterSensitiveLog,
  GetAgentVersionCommand: () => GetAgentVersionCommand,
  GetAgentVersionResponseFilterSensitiveLog: () => GetAgentVersionResponseFilterSensitiveLog,
  GetDataSourceCommand: () => GetDataSourceCommand,
  GetIngestionJobCommand: () => GetIngestionJobCommand,
  GetKnowledgeBaseCommand: () => GetKnowledgeBaseCommand,
  IngestionJobFilterAttribute: () => IngestionJobFilterAttribute,
  IngestionJobFilterOperator: () => IngestionJobFilterOperator,
  IngestionJobSortByAttribute: () => IngestionJobSortByAttribute,
  IngestionJobStatus: () => IngestionJobStatus,
  InternalServerException: () => InternalServerException,
  KnowledgeBaseState: () => KnowledgeBaseState,
  KnowledgeBaseStatus: () => KnowledgeBaseStatus,
  KnowledgeBaseStorageType: () => KnowledgeBaseStorageType,
  KnowledgeBaseType: () => KnowledgeBaseType,
  ListAgentActionGroupsCommand: () => ListAgentActionGroupsCommand,
  ListAgentAliasesCommand: () => ListAgentAliasesCommand,
  ListAgentKnowledgeBasesCommand: () => ListAgentKnowledgeBasesCommand,
  ListAgentVersionsCommand: () => ListAgentVersionsCommand,
  ListAgentsCommand: () => ListAgentsCommand,
  ListDataSourcesCommand: () => ListDataSourcesCommand,
  ListIngestionJobsCommand: () => ListIngestionJobsCommand,
  ListKnowledgeBasesCommand: () => ListKnowledgeBasesCommand,
  ListTagsForResourceCommand: () => ListTagsForResourceCommand,
  PrepareAgentCommand: () => PrepareAgentCommand,
  PromptOverrideConfigurationFilterSensitiveLog: () => PromptOverrideConfigurationFilterSensitiveLog,
  PromptState: () => PromptState,
  PromptType: () => PromptType,
  ResourceNotFoundException: () => ResourceNotFoundException,
  ServiceQuotaExceededException: () => ServiceQuotaExceededException,
  SortOrder: () => SortOrder,
  StartIngestionJobCommand: () => StartIngestionJobCommand,
  TagResourceCommand: () => TagResourceCommand,
  ThrottlingException: () => ThrottlingException,
  UntagResourceCommand: () => UntagResourceCommand,
  UpdateAgentActionGroupCommand: () => UpdateAgentActionGroupCommand,
  UpdateAgentActionGroupRequestFilterSensitiveLog: () => UpdateAgentActionGroupRequestFilterSensitiveLog,
  UpdateAgentActionGroupResponseFilterSensitiveLog: () => UpdateAgentActionGroupResponseFilterSensitiveLog,
  UpdateAgentAliasCommand: () => UpdateAgentAliasCommand,
  UpdateAgentCommand: () => UpdateAgentCommand,
  UpdateAgentKnowledgeBaseCommand: () => UpdateAgentKnowledgeBaseCommand,
  UpdateAgentRequestFilterSensitiveLog: () => UpdateAgentRequestFilterSensitiveLog,
  UpdateAgentResponseFilterSensitiveLog: () => UpdateAgentResponseFilterSensitiveLog,
  UpdateDataSourceCommand: () => UpdateDataSourceCommand,
  UpdateKnowledgeBaseCommand: () => UpdateKnowledgeBaseCommand,
  ValidationException: () => ValidationException,
  __Client: () => import_smithy_client.Client,
  paginateListAgentActionGroups: () => paginateListAgentActionGroups,
  paginateListAgentAliases: () => paginateListAgentAliases,
  paginateListAgentKnowledgeBases: () => paginateListAgentKnowledgeBases,
  paginateListAgentVersions: () => paginateListAgentVersions,
  paginateListAgents: () => paginateListAgents,
  paginateListDataSources: () => paginateListDataSources,
  paginateListIngestionJobs: () => paginateListIngestionJobs,
  paginateListKnowledgeBases: () => paginateListKnowledgeBases
});
module.exports = __toCommonJS(src_exports);

// src/BedrockAgentClient.ts
var import_middleware_host_header = require("@aws-sdk/middleware-host-header");
var import_middleware_logger = require("@aws-sdk/middleware-logger");
var import_middleware_recursion_detection = require("@aws-sdk/middleware-recursion-detection");
var import_middleware_user_agent = require("@aws-sdk/middleware-user-agent");
var import_config_resolver = require("@smithy/config-resolver");
var import_core = require("@smithy/core");
var import_middleware_content_length = require("@smithy/middleware-content-length");
var import_middleware_endpoint = require("@smithy/middleware-endpoint");
var import_middleware_retry = require("@smithy/middleware-retry");

var import_httpAuthSchemeProvider = require("./auth/httpAuthSchemeProvider");

// src/endpoint/EndpointParameters.ts
var resolveClientEndpointParameters = /* @__PURE__ */ __name((options) => {
  return {
    ...options,
    useDualstackEndpoint: options.useDualstackEndpoint ?? false,
    useFipsEndpoint: options.useFipsEndpoint ?? false,
    defaultSigningName: "bedrock"
  };
}, "resolveClientEndpointParameters");
var commonParams = {
  UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
  Endpoint: { type: "builtInParams", name: "endpoint" },
  Region: { type: "builtInParams", name: "region" },
  UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" }
};

// src/BedrockAgentClient.ts
var import_runtimeConfig = require("././runtimeConfig");

// src/runtimeExtensions.ts
var import_region_config_resolver = require("@aws-sdk/region-config-resolver");
var import_protocol_http = require("@smithy/protocol-http");
var import_smithy_client = require("@smithy/smithy-client");

// src/auth/httpAuthExtensionConfiguration.ts
var getHttpAuthExtensionConfiguration = /* @__PURE__ */ __name((runtimeConfig) => {
  const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
  let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
  let _credentials = runtimeConfig.credentials;
  return {
    setHttpAuthScheme(httpAuthScheme) {
      const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
      if (index === -1) {
        _httpAuthSchemes.push(httpAuthScheme);
      } else {
        _httpAuthSchemes.splice(index, 1, httpAuthScheme);
      }
    },
    httpAuthSchemes() {
      return _httpAuthSchemes;
    },
    setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
      _httpAuthSchemeProvider = httpAuthSchemeProvider;
    },
    httpAuthSchemeProvider() {
      return _httpAuthSchemeProvider;
    },
    setCredentials(credentials) {
      _credentials = credentials;
    },
    credentials() {
      return _credentials;
    }
  };
}, "getHttpAuthExtensionConfiguration");
var resolveHttpAuthRuntimeConfig = /* @__PURE__ */ __name((config) => {
  return {
    httpAuthSchemes: config.httpAuthSchemes(),
    httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
    credentials: config.credentials()
  };
}, "resolveHttpAuthRuntimeConfig");

// src/runtimeExtensions.ts
var asPartial = /* @__PURE__ */ __name((t) => t, "asPartial");
var resolveRuntimeExtensions = /* @__PURE__ */ __name((runtimeConfig, extensions) => {
  const extensionConfiguration = {
    ...asPartial((0, import_region_config_resolver.getAwsRegionExtensionConfiguration)(runtimeConfig)),
    ...asPartial((0, import_smithy_client.getDefaultExtensionConfiguration)(runtimeConfig)),
    ...asPartial((0, import_protocol_http.getHttpHandlerExtensionConfiguration)(runtimeConfig)),
    ...asPartial(getHttpAuthExtensionConfiguration(runtimeConfig))
  };
  extensions.forEach((extension) => extension.configure(extensionConfiguration));
  return {
    ...runtimeConfig,
    ...(0, import_region_config_resolver.resolveAwsRegionExtensionConfiguration)(extensionConfiguration),
    ...(0, import_smithy_client.resolveDefaultRuntimeConfig)(extensionConfiguration),
    ...(0, import_protocol_http.resolveHttpHandlerRuntimeConfig)(extensionConfiguration),
    ...resolveHttpAuthRuntimeConfig(extensionConfiguration)
  };
}, "resolveRuntimeExtensions");

// src/BedrockAgentClient.ts
var _BedrockAgentClient = class _BedrockAgentClient extends import_smithy_client.Client {
  constructor(...[configuration]) {
    const _config_0 = (0, import_runtimeConfig.getRuntimeConfig)(configuration || {});
    const _config_1 = resolveClientEndpointParameters(_config_0);
    const _config_2 = (0, import_config_resolver.resolveRegionConfig)(_config_1);
    const _config_3 = (0, import_middleware_endpoint.resolveEndpointConfig)(_config_2);
    const _config_4 = (0, import_middleware_retry.resolveRetryConfig)(_config_3);
    const _config_5 = (0, import_middleware_host_header.resolveHostHeaderConfig)(_config_4);
    const _config_6 = (0, import_middleware_user_agent.resolveUserAgentConfig)(_config_5);
    const _config_7 = (0, import_httpAuthSchemeProvider.resolveHttpAuthSchemeConfig)(_config_6);
    const _config_8 = resolveRuntimeExtensions(_config_7, (configuration == null ? void 0 : configuration.extensions) || []);
    super(_config_8);
    this.config = _config_8;
    this.middlewareStack.use((0, import_middleware_retry.getRetryPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_content_length.getContentLengthPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_host_header.getHostHeaderPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_logger.getLoggerPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_recursion_detection.getRecursionDetectionPlugin)(this.config));
    this.middlewareStack.use((0, import_middleware_user_agent.getUserAgentPlugin)(this.config));
    this.middlewareStack.use(
      (0, import_core.getHttpAuthSchemeEndpointRuleSetPlugin)(this.config, {
        httpAuthSchemeParametersProvider: this.getDefaultHttpAuthSchemeParametersProvider(),
        identityProviderConfigProvider: this.getIdentityProviderConfigProvider()
      })
    );
    this.middlewareStack.use((0, import_core.getHttpSigningPlugin)(this.config));
  }
  /**
   * Destroy underlying resources, like sockets. It's usually not necessary to do this.
   * However in Node.js, it's best to explicitly shut down the client's agent when it is no longer needed.
   * Otherwise, sockets might stay open for quite a long time before the server terminates them.
   */
  destroy() {
    super.destroy();
  }
  getDefaultHttpAuthSchemeParametersProvider() {
    return import_httpAuthSchemeProvider.defaultBedrockAgentHttpAuthSchemeParametersProvider;
  }
  getIdentityProviderConfigProvider() {
    return async (config) => new import_core.DefaultIdentityProviderConfig({
      "aws.auth#sigv4": config.credentials
    });
  }
};
__name(_BedrockAgentClient, "BedrockAgentClient");
var BedrockAgentClient = _BedrockAgentClient;

// src/BedrockAgent.ts


// src/commands/AssociateAgentKnowledgeBaseCommand.ts

var import_middleware_serde = require("@smithy/middleware-serde");

var import_types = require("@smithy/types");

// src/protocols/Aws_restJson1.ts
var import_core2 = require("@aws-sdk/core");


var import_uuid = require("uuid");

// src/models/BedrockAgentServiceException.ts

var _BedrockAgentServiceException = class _BedrockAgentServiceException extends import_smithy_client.ServiceException {
  /**
   * @internal
   */
  constructor(options) {
    super(options);
    Object.setPrototypeOf(this, _BedrockAgentServiceException.prototype);
  }
};
__name(_BedrockAgentServiceException, "BedrockAgentServiceException");
var BedrockAgentServiceException = _BedrockAgentServiceException;

// src/models/models_0.ts

var _AccessDeniedException = class _AccessDeniedException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "AccessDeniedException",
      $fault: "client",
      ...opts
    });
    this.name = "AccessDeniedException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _AccessDeniedException.prototype);
  }
};
__name(_AccessDeniedException, "AccessDeniedException");
var AccessDeniedException = _AccessDeniedException;
var ActionGroupExecutor;
((ActionGroupExecutor3) => {
  ActionGroupExecutor3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.lambda !== void 0)
      return visitor.lambda(value.lambda);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(ActionGroupExecutor || (ActionGroupExecutor = {}));
var _ConflictException = class _ConflictException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ConflictException",
      $fault: "client",
      ...opts
    });
    this.name = "ConflictException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ConflictException.prototype);
  }
};
__name(_ConflictException, "ConflictException");
var ConflictException = _ConflictException;
var ActionGroupState = {
  DISABLED: "DISABLED",
  ENABLED: "ENABLED"
};
var APISchema;
((APISchema3) => {
  APISchema3.visit = /* @__PURE__ */ __name((value, visitor) => {
    if (value.s3 !== void 0)
      return visitor.s3(value.s3);
    if (value.payload !== void 0)
      return visitor.payload(value.payload);
    return visitor._(value.$unknown[0], value.$unknown[1]);
  }, "visit");
})(APISchema || (APISchema = {}));
var ActionGroupSignature = {
  AMAZON_USERINPUT: "AMAZON.UserInput"
};
var _InternalServerException = class _InternalServerException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "InternalServerException",
      $fault: "server",
      ...opts
    });
    this.name = "InternalServerException";
    this.$fault = "server";
    Object.setPrototypeOf(this, _InternalServerException.prototype);
  }
};
__name(_InternalServerException, "InternalServerException");
var InternalServerException = _InternalServerException;
var _ResourceNotFoundException = class _ResourceNotFoundException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ResourceNotFoundException",
      $fault: "client",
      ...opts
    });
    this.name = "ResourceNotFoundException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ResourceNotFoundException.prototype);
  }
};
__name(_ResourceNotFoundException, "ResourceNotFoundException");
var ResourceNotFoundException = _ResourceNotFoundException;
var _ServiceQuotaExceededException = class _ServiceQuotaExceededException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ServiceQuotaExceededException",
      $fault: "client",
      ...opts
    });
    this.name = "ServiceQuotaExceededException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ServiceQuotaExceededException.prototype);
  }
};
__name(_ServiceQuotaExceededException, "ServiceQuotaExceededException");
var ServiceQuotaExceededException = _ServiceQuotaExceededException;
var _ThrottlingException = class _ThrottlingException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ThrottlingException",
      $fault: "client",
      ...opts
    });
    this.name = "ThrottlingException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ThrottlingException.prototype);
  }
};
__name(_ThrottlingException, "ThrottlingException");
var ThrottlingException = _ThrottlingException;
var _ValidationException = class _ValidationException extends BedrockAgentServiceException {
  /**
   * @internal
   */
  constructor(opts) {
    super({
      name: "ValidationException",
      $fault: "client",
      ...opts
    });
    this.name = "ValidationException";
    this.$fault = "client";
    Object.setPrototypeOf(this, _ValidationException.prototype);
    this.fieldList = opts.fieldList;
  }
};
__name(_ValidationException, "ValidationException");
var ValidationException = _ValidationException;
var AgentStatus = {
  CREATING: "CREATING",
  DELETING: "DELETING",
  FAILED: "FAILED",
  NOT_PREPARED: "NOT_PREPARED",
  PREPARED: "PREPARED",
  PREPARING: "PREPARING",
  UPDATING: "UPDATING",
  VERSIONING: "VERSIONING"
};
var CreationMode = {
  DEFAULT: "DEFAULT",
  OVERRIDDEN: "OVERRIDDEN"
};
var PromptState = {
  DISABLED: "DISABLED",
  ENABLED: "ENABLED"
};
var PromptType = {
  KNOWLEDGE_BASE_RESPONSE_GENERATION: "KNOWLEDGE_BASE_RESPONSE_GENERATION",
  ORCHESTRATION: "ORCHESTRATION",
  POST_PROCESSING: "POST_PROCESSING",
  PRE_PROCESSING: "PRE_PROCESSING"
};
var AgentAliasStatus = {
  CREATING: "CREATING",
  DELETING: "DELETING",
  FAILED: "FAILED",
  PREPARED: "PREPARED",
  UPDATING: "UPDATING"
};
var KnowledgeBaseState = {
  DISABLED: "DISABLED",
  ENABLED: "ENABLED"
};
var DataSourceType = {
  S3: "S3"
};
var ChunkingStrategy = {
  FIXED_SIZE: "FIXED_SIZE",
  NONE: "NONE"
};
var DataSourceStatus = {
  AVAILABLE: "AVAILABLE",
  DELETING: "DELETING"
};
var IngestionJobStatus = {
  COMPLETE: "COMPLETE",
  FAILED: "FAILED",
  IN_PROGRESS: "IN_PROGRESS",
  STARTING: "STARTING"
};
var IngestionJobFilterAttribute = {
  STATUS: "STATUS"
};
var IngestionJobFilterOperator = {
  EQ: "EQ"
};
var IngestionJobSortByAttribute = {
  STARTED_AT: "STARTED_AT",
  STATUS: "STATUS"
};
var SortOrder = {
  ASCENDING: "ASCENDING",
  DESCENDING: "DESCENDING"
};
var KnowledgeBaseType = {
  VECTOR: "VECTOR"
};
var KnowledgeBaseStorageType = {
  OPENSEARCH_SERVERLESS: "OPENSEARCH_SERVERLESS",
  PINECONE: "PINECONE",
  RDS: "RDS",
  REDIS_ENTERPRISE_CLOUD: "REDIS_ENTERPRISE_CLOUD"
};
var KnowledgeBaseStatus = {
  ACTIVE: "ACTIVE",
  CREATING: "CREATING",
  DELETING: "DELETING",
  FAILED: "FAILED",
  UPDATING: "UPDATING"
};
var APISchemaFilterSensitiveLog = /* @__PURE__ */ __name((obj) => {
  if (obj.s3 !== void 0)
    return { s3: obj.s3 };
  if (obj.payload !== void 0)
    return { payload: import_smithy_client.SENSITIVE_STRING };
  if (obj.$unknown !== void 0)
    return { [obj.$unknown[0]]: "UNKNOWN" };
}, "APISchemaFilterSensitiveLog");
var CreateAgentActionGroupRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupExecutor && { actionGroupExecutor: obj.actionGroupExecutor },
  ...obj.apiSchema && { apiSchema: APISchemaFilterSensitiveLog(obj.apiSchema) }
}), "CreateAgentActionGroupRequestFilterSensitiveLog");
var AgentActionGroupFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupExecutor && { actionGroupExecutor: obj.actionGroupExecutor },
  ...obj.apiSchema && { apiSchema: APISchemaFilterSensitiveLog(obj.apiSchema) }
}), "AgentActionGroupFilterSensitiveLog");
var CreateAgentActionGroupResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agentActionGroup && { agentActionGroup: AgentActionGroupFilterSensitiveLog(obj.agentActionGroup) }
}), "CreateAgentActionGroupResponseFilterSensitiveLog");
var GetAgentActionGroupResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agentActionGroup && { agentActionGroup: AgentActionGroupFilterSensitiveLog(obj.agentActionGroup) }
}), "GetAgentActionGroupResponseFilterSensitiveLog");
var UpdateAgentActionGroupRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.actionGroupExecutor && { actionGroupExecutor: obj.actionGroupExecutor },
  ...obj.apiSchema && { apiSchema: APISchemaFilterSensitiveLog(obj.apiSchema) }
}), "UpdateAgentActionGroupRequestFilterSensitiveLog");
var UpdateAgentActionGroupResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agentActionGroup && { agentActionGroup: AgentActionGroupFilterSensitiveLog(obj.agentActionGroup) }
}), "UpdateAgentActionGroupResponseFilterSensitiveLog");
var PromptOverrideConfigurationFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj
}), "PromptOverrideConfigurationFilterSensitiveLog");
var AgentFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.instruction && { instruction: import_smithy_client.SENSITIVE_STRING },
  ...obj.promptOverrideConfiguration && { promptOverrideConfiguration: import_smithy_client.SENSITIVE_STRING }
}), "AgentFilterSensitiveLog");
var CreateAgentRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.instruction && { instruction: import_smithy_client.SENSITIVE_STRING },
  ...obj.promptOverrideConfiguration && { promptOverrideConfiguration: import_smithy_client.SENSITIVE_STRING }
}), "CreateAgentRequestFilterSensitiveLog");
var CreateAgentResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agent && { agent: AgentFilterSensitiveLog(obj.agent) }
}), "CreateAgentResponseFilterSensitiveLog");
var GetAgentResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agent && { agent: AgentFilterSensitiveLog(obj.agent) }
}), "GetAgentResponseFilterSensitiveLog");
var UpdateAgentRequestFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.instruction && { instruction: import_smithy_client.SENSITIVE_STRING },
  ...obj.promptOverrideConfiguration && { promptOverrideConfiguration: import_smithy_client.SENSITIVE_STRING }
}), "UpdateAgentRequestFilterSensitiveLog");
var UpdateAgentResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agent && { agent: AgentFilterSensitiveLog(obj.agent) }
}), "UpdateAgentResponseFilterSensitiveLog");
var AgentVersionFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.instruction && { instruction: import_smithy_client.SENSITIVE_STRING },
  ...obj.promptOverrideConfiguration && { promptOverrideConfiguration: import_smithy_client.SENSITIVE_STRING }
}), "AgentVersionFilterSensitiveLog");
var GetAgentVersionResponseFilterSensitiveLog = /* @__PURE__ */ __name((obj) => ({
  ...obj,
  ...obj.agentVersion && { agentVersion: AgentVersionFilterSensitiveLog(obj.agentVersion) }
}), "GetAgentVersionResponseFilterSensitiveLog");

// src/protocols/Aws_restJson1.ts
var se_AssociateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/knowledgebases");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      description: [],
      knowledgeBaseId: [],
      knowledgeBaseState: []
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_AssociateAgentKnowledgeBaseCommand");
var se_CreateAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents");
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      agentName: [],
      agentResourceRoleArn: [],
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      customerEncryptionKeyArn: [],
      description: [],
      foundationModel: [],
      idleSessionTTLInSeconds: [],
      instruction: [],
      promptOverrideConfiguration: (_) => se_PromptOverrideConfiguration(_, context),
      tags: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_CreateAgentCommand");
var se_CreateAgentActionGroupCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/actiongroups");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      actionGroupExecutor: (_) => (0, import_smithy_client._json)(_),
      actionGroupName: [],
      actionGroupState: [],
      apiSchema: (_) => (0, import_smithy_client._json)(_),
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      description: [],
      parentActionGroupSignature: []
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_CreateAgentActionGroupCommand");
var se_CreateAgentAliasCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentaliases");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      agentAliasName: [],
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      description: [],
      routingConfiguration: (_) => (0, import_smithy_client._json)(_),
      tags: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_CreateAgentAliasCommand");
var se_CreateDataSourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      dataSourceConfiguration: (_) => (0, import_smithy_client._json)(_),
      description: [],
      name: [],
      serverSideEncryptionConfiguration: (_) => (0, import_smithy_client._json)(_),
      vectorIngestionConfiguration: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_CreateDataSourceCommand");
var se_CreateKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases");
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      description: [],
      knowledgeBaseConfiguration: (_) => (0, import_smithy_client._json)(_),
      name: [],
      roleArn: [],
      storageConfiguration: (_) => (0, import_smithy_client._json)(_),
      tags: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_CreateKnowledgeBaseCommand");
var se_DeleteAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  const query = (0, import_smithy_client.map)({
    [_sRIUC]: [() => input.skipResourceInUseCheck !== void 0, () => input[_sRIUC].toString()]
  });
  let body;
  b.m("DELETE").h(headers).q(query).b(body);
  return b.build();
}, "se_DeleteAgentCommand");
var se_DeleteAgentActionGroupCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/actiongroups/{actionGroupId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("actionGroupId", () => input.actionGroupId, "{actionGroupId}", false);
  const query = (0, import_smithy_client.map)({
    [_sRIUC]: [() => input.skipResourceInUseCheck !== void 0, () => input[_sRIUC].toString()]
  });
  let body;
  b.m("DELETE").h(headers).q(query).b(body);
  return b.build();
}, "se_DeleteAgentActionGroupCommand");
var se_DeleteAgentAliasCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentaliases/{agentAliasId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentAliasId", () => input.agentAliasId, "{agentAliasId}", false);
  let body;
  b.m("DELETE").h(headers).b(body);
  return b.build();
}, "se_DeleteAgentAliasCommand");
var se_DeleteAgentVersionCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  const query = (0, import_smithy_client.map)({
    [_sRIUC]: [() => input.skipResourceInUseCheck !== void 0, () => input[_sRIUC].toString()]
  });
  let body;
  b.m("DELETE").h(headers).q(query).b(body);
  return b.build();
}, "se_DeleteAgentVersionCommand");
var se_DeleteDataSourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  let body;
  b.m("DELETE").h(headers).b(body);
  return b.build();
}, "se_DeleteDataSourceCommand");
var se_DeleteKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/knowledgebases/{knowledgeBaseId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  b.m("DELETE").h(headers).b(body);
  return b.build();
}, "se_DeleteKnowledgeBaseCommand");
var se_DisassociateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/knowledgebases/{knowledgeBaseId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  b.m("DELETE").h(headers).b(body);
  return b.build();
}, "se_DisassociateAgentKnowledgeBaseCommand");
var se_GetAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetAgentCommand");
var se_GetAgentActionGroupCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/actiongroups/{actionGroupId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("actionGroupId", () => input.actionGroupId, "{actionGroupId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetAgentActionGroupCommand");
var se_GetAgentAliasCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentaliases/{agentAliasId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentAliasId", () => input.agentAliasId, "{agentAliasId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetAgentAliasCommand");
var se_GetAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/knowledgebases/{knowledgeBaseId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetAgentKnowledgeBaseCommand");
var se_GetAgentVersionCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}/agentversions/{agentVersion}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetAgentVersionCommand");
var se_GetDataSourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetDataSourceCommand");
var se_GetIngestionJobCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}/ingestionjobs/{ingestionJobId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  b.p("ingestionJobId", () => input.ingestionJobId, "{ingestionJobId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetIngestionJobCommand");
var se_GetKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/knowledgebases/{knowledgeBaseId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_GetKnowledgeBaseCommand");
var se_ListAgentActionGroupsCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/actiongroups");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListAgentActionGroupsCommand");
var se_ListAgentAliasesCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentaliases");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListAgentAliasesCommand");
var se_ListAgentKnowledgeBasesCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/knowledgebases");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListAgentKnowledgeBasesCommand");
var se_ListAgentsCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents");
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListAgentsCommand");
var se_ListAgentVersionsCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListAgentVersionsCommand");
var se_ListDataSourcesCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListDataSourcesCommand");
var se_ListIngestionJobsCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}/ingestionjobs");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      filters: (_) => (0, import_smithy_client._json)(_),
      maxResults: [],
      nextToken: [],
      sortBy: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListIngestionJobsCommand");
var se_ListKnowledgeBasesCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases");
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      maxResults: [],
      nextToken: []
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_ListKnowledgeBasesCommand");
var se_ListTagsForResourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/tags/{resourceArn}");
  b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
  let body;
  b.m("GET").h(headers).b(body);
  return b.build();
}, "se_ListTagsForResourceCommand");
var se_PrepareAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/agents/{agentId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_PrepareAgentCommand");
var se_StartIngestionJobCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}/ingestionjobs");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      clientToken: [true, (_) => _ ?? (0, import_uuid.v4)()],
      description: []
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_StartIngestionJobCommand");
var se_TagResourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/tags/{resourceArn}");
  b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      tags: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("POST").h(headers).b(body);
  return b.build();
}, "se_TagResourceCommand");
var se_UntagResourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {};
  b.bp("/tags/{resourceArn}");
  b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
  const query = (0, import_smithy_client.map)({
    [_tK]: [
      (0, import_smithy_client.expectNonNull)(input.tagKeys, `tagKeys`) != null,
      () => (input[_tK] || []).map((_entry) => _entry)
    ]
  });
  let body;
  b.m("DELETE").h(headers).q(query).b(body);
  return b.build();
}, "se_UntagResourceCommand");
var se_UpdateAgentCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      agentName: [],
      agentResourceRoleArn: [],
      customerEncryptionKeyArn: [],
      description: [],
      foundationModel: [],
      idleSessionTTLInSeconds: [],
      instruction: [],
      promptOverrideConfiguration: (_) => se_PromptOverrideConfiguration(_, context)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateAgentCommand");
var se_UpdateAgentActionGroupCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/actiongroups/{actionGroupId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("actionGroupId", () => input.actionGroupId, "{actionGroupId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      actionGroupExecutor: (_) => (0, import_smithy_client._json)(_),
      actionGroupName: [],
      actionGroupState: [],
      apiSchema: (_) => (0, import_smithy_client._json)(_),
      description: [],
      parentActionGroupSignature: []
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateAgentActionGroupCommand");
var se_UpdateAgentAliasCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentaliases/{agentAliasId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentAliasId", () => input.agentAliasId, "{agentAliasId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      agentAliasName: [],
      description: [],
      routingConfiguration: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateAgentAliasCommand");
var se_UpdateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/agents/{agentId}/agentversions/{agentVersion}/knowledgebases/{knowledgeBaseId}");
  b.p("agentId", () => input.agentId, "{agentId}", false);
  b.p("agentVersion", () => input.agentVersion, "{agentVersion}", false);
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      description: [],
      knowledgeBaseState: []
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateAgentKnowledgeBaseCommand");
var se_UpdateDataSourceCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}/datasources/{dataSourceId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  b.p("dataSourceId", () => input.dataSourceId, "{dataSourceId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      dataSourceConfiguration: (_) => (0, import_smithy_client._json)(_),
      description: [],
      name: [],
      serverSideEncryptionConfiguration: (_) => (0, import_smithy_client._json)(_),
      vectorIngestionConfiguration: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateDataSourceCommand");
var se_UpdateKnowledgeBaseCommand = /* @__PURE__ */ __name(async (input, context) => {
  const b = (0, import_core.requestBuilder)(input, context);
  const headers = {
    "content-type": "application/json"
  };
  b.bp("/knowledgebases/{knowledgeBaseId}");
  b.p("knowledgeBaseId", () => input.knowledgeBaseId, "{knowledgeBaseId}", false);
  let body;
  body = JSON.stringify(
    (0, import_smithy_client.take)(input, {
      description: [],
      knowledgeBaseConfiguration: (_) => (0, import_smithy_client._json)(_),
      name: [],
      roleArn: [],
      storageConfiguration: (_) => (0, import_smithy_client._json)(_)
    })
  );
  b.m("PUT").h(headers).b(body);
  return b.build();
}, "se_UpdateKnowledgeBaseCommand");
var de_AssociateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentKnowledgeBase: (_) => de_AgentKnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_AssociateAgentKnowledgeBaseCommand");
var de_CreateAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agent: (_) => de_Agent(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_CreateAgentCommand");
var de_CreateAgentActionGroupCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentActionGroup: (_) => de_AgentActionGroup(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_CreateAgentActionGroupCommand");
var de_CreateAgentAliasCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentAlias: (_) => de_AgentAlias(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_CreateAgentAliasCommand");
var de_CreateDataSourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    dataSource: (_) => de_DataSource(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_CreateDataSourceCommand");
var de_CreateKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    knowledgeBase: (_) => de_KnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_CreateKnowledgeBaseCommand");
var de_DeleteAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentId: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_DeleteAgentCommand");
var de_DeleteAgentActionGroupCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 204 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  await (0, import_smithy_client.collectBody)(output.body, context);
  return contents;
}, "de_DeleteAgentActionGroupCommand");
var de_DeleteAgentAliasCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentAliasId: import_smithy_client.expectString,
    agentAliasStatus: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_DeleteAgentAliasCommand");
var de_DeleteAgentVersionCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentId: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_DeleteAgentVersionCommand");
var de_DeleteDataSourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    dataSourceId: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    status: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_DeleteDataSourceCommand");
var de_DeleteKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    knowledgeBaseId: import_smithy_client.expectString,
    status: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_DeleteKnowledgeBaseCommand");
var de_DisassociateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 204 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  await (0, import_smithy_client.collectBody)(output.body, context);
  return contents;
}, "de_DisassociateAgentKnowledgeBaseCommand");
var de_GetAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agent: (_) => de_Agent(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetAgentCommand");
var de_GetAgentActionGroupCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentActionGroup: (_) => de_AgentActionGroup(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetAgentActionGroupCommand");
var de_GetAgentAliasCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentAlias: (_) => de_AgentAlias(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetAgentAliasCommand");
var de_GetAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentKnowledgeBase: (_) => de_AgentKnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetAgentKnowledgeBaseCommand");
var de_GetAgentVersionCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentVersion: (_) => de_AgentVersion(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetAgentVersionCommand");
var de_GetDataSourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    dataSource: (_) => de_DataSource(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetDataSourceCommand");
var de_GetIngestionJobCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    ingestionJob: (_) => de_IngestionJob(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetIngestionJobCommand");
var de_GetKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    knowledgeBase: (_) => de_KnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_GetKnowledgeBaseCommand");
var de_ListAgentActionGroupsCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    actionGroupSummaries: (_) => de_ActionGroupSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListAgentActionGroupsCommand");
var de_ListAgentAliasesCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentAliasSummaries: (_) => de_AgentAliasSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListAgentAliasesCommand");
var de_ListAgentKnowledgeBasesCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentKnowledgeBaseSummaries: (_) => de_AgentKnowledgeBaseSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListAgentKnowledgeBasesCommand");
var de_ListAgentsCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentSummaries: (_) => de_AgentSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListAgentsCommand");
var de_ListAgentVersionsCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentVersionSummaries: (_) => de_AgentVersionSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListAgentVersionsCommand");
var de_ListDataSourcesCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    dataSourceSummaries: (_) => de_DataSourceSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListDataSourcesCommand");
var de_ListIngestionJobsCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    ingestionJobSummaries: (_) => de_IngestionJobSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListIngestionJobsCommand");
var de_ListKnowledgeBasesCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    knowledgeBaseSummaries: (_) => de_KnowledgeBaseSummaries(_, context),
    nextToken: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListKnowledgeBasesCommand");
var de_ListTagsForResourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    tags: import_smithy_client._json
  });
  Object.assign(contents, doc);
  return contents;
}, "de_ListTagsForResourceCommand");
var de_PrepareAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentId: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString,
    preparedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
  Object.assign(contents, doc);
  return contents;
}, "de_PrepareAgentCommand");
var de_StartIngestionJobCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    ingestionJob: (_) => de_IngestionJob(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_StartIngestionJobCommand");
var de_TagResourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  await (0, import_smithy_client.collectBody)(output.body, context);
  return contents;
}, "de_TagResourceCommand");
var de_UntagResourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  await (0, import_smithy_client.collectBody)(output.body, context);
  return contents;
}, "de_UntagResourceCommand");
var de_UpdateAgentCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agent: (_) => de_Agent(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateAgentCommand");
var de_UpdateAgentActionGroupCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentActionGroup: (_) => de_AgentActionGroup(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateAgentActionGroupCommand");
var de_UpdateAgentAliasCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentAlias: (_) => de_AgentAlias(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateAgentAliasCommand");
var de_UpdateAgentKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    agentKnowledgeBase: (_) => de_AgentKnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateAgentKnowledgeBaseCommand");
var de_UpdateDataSourceCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 200 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    dataSource: (_) => de_DataSource(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateDataSourceCommand");
var de_UpdateKnowledgeBaseCommand = /* @__PURE__ */ __name(async (output, context) => {
  if (output.statusCode !== 202 && output.statusCode >= 300) {
    return de_CommandError(output, context);
  }
  const contents = (0, import_smithy_client.map)({
    $metadata: deserializeMetadata(output)
  });
  const data = (0, import_smithy_client.expectNonNull)((0, import_smithy_client.expectObject)(await parseBody(output.body, context)), "body");
  const doc = (0, import_smithy_client.take)(data, {
    knowledgeBase: (_) => de_KnowledgeBase(_, context)
  });
  Object.assign(contents, doc);
  return contents;
}, "de_UpdateKnowledgeBaseCommand");
var de_CommandError = /* @__PURE__ */ __name(async (output, context) => {
  const parsedOutput = {
    ...output,
    body: await parseErrorBody(output.body, context)
  };
  const errorCode = loadRestJsonErrorCode(output, parsedOutput.body);
  switch (errorCode) {
    case "AccessDeniedException":
    case "com.amazonaws.bedrockagent#AccessDeniedException":
      throw await de_AccessDeniedExceptionRes(parsedOutput, context);
    case "ConflictException":
    case "com.amazonaws.bedrockagent#ConflictException":
      throw await de_ConflictExceptionRes(parsedOutput, context);
    case "InternalServerException":
    case "com.amazonaws.bedrockagent#InternalServerException":
      throw await de_InternalServerExceptionRes(parsedOutput, context);
    case "ResourceNotFoundException":
    case "com.amazonaws.bedrockagent#ResourceNotFoundException":
      throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
    case "ServiceQuotaExceededException":
    case "com.amazonaws.bedrockagent#ServiceQuotaExceededException":
      throw await de_ServiceQuotaExceededExceptionRes(parsedOutput, context);
    case "ThrottlingException":
    case "com.amazonaws.bedrockagent#ThrottlingException":
      throw await de_ThrottlingExceptionRes(parsedOutput, context);
    case "ValidationException":
    case "com.amazonaws.bedrockagent#ValidationException":
      throw await de_ValidationExceptionRes(parsedOutput, context);
    default:
      const parsedBody = parsedOutput.body;
      return throwDefaultError({
        output,
        parsedBody,
        errorCode
      });
  }
}, "de_CommandError");
var throwDefaultError = (0, import_smithy_client.withBaseException)(BedrockAgentServiceException);
var de_AccessDeniedExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new AccessDeniedException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_AccessDeniedExceptionRes");
var de_ConflictExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ConflictException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ConflictExceptionRes");
var de_InternalServerExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new InternalServerException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_InternalServerExceptionRes");
var de_ResourceNotFoundExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ResourceNotFoundException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ResourceNotFoundExceptionRes");
var de_ServiceQuotaExceededExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ServiceQuotaExceededException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ServiceQuotaExceededExceptionRes");
var de_ThrottlingExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ThrottlingException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ThrottlingExceptionRes");
var de_ValidationExceptionRes = /* @__PURE__ */ __name(async (parsedOutput, context) => {
  const contents = (0, import_smithy_client.map)({});
  const data = parsedOutput.body;
  const doc = (0, import_smithy_client.take)(data, {
    fieldList: import_smithy_client._json,
    message: import_smithy_client.expectString
  });
  Object.assign(contents, doc);
  const exception = new ValidationException({
    $metadata: deserializeMetadata(parsedOutput),
    ...contents
  });
  return (0, import_smithy_client.decorateServiceException)(exception, parsedOutput.body);
}, "de_ValidationExceptionRes");
var se_InferenceConfiguration = /* @__PURE__ */ __name((input, context) => {
  return (0, import_smithy_client.take)(input, {
    maximumLength: [],
    stopSequences: import_smithy_client._json,
    temperature: import_smithy_client.serializeFloat,
    topK: [],
    topP: import_smithy_client.serializeFloat
  });
}, "se_InferenceConfiguration");
var se_PromptConfiguration = /* @__PURE__ */ __name((input, context) => {
  return (0, import_smithy_client.take)(input, {
    basePromptTemplate: [],
    inferenceConfiguration: (_) => se_InferenceConfiguration(_, context),
    parserMode: [],
    promptCreationMode: [],
    promptState: [],
    promptType: []
  });
}, "se_PromptConfiguration");
var se_PromptConfigurations = /* @__PURE__ */ __name((input, context) => {
  return input.filter((e) => e != null).map((entry) => {
    return se_PromptConfiguration(entry, context);
  });
}, "se_PromptConfigurations");
var se_PromptOverrideConfiguration = /* @__PURE__ */ __name((input, context) => {
  return (0, import_smithy_client.take)(input, {
    overrideLambda: [],
    promptConfigurations: (_) => se_PromptConfigurations(_, context)
  });
}, "se_PromptOverrideConfiguration");
var de_ActionGroupSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_ActionGroupSummary(entry, context);
  });
  return retVal;
}, "de_ActionGroupSummaries");
var de_ActionGroupSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    actionGroupId: import_smithy_client.expectString,
    actionGroupName: import_smithy_client.expectString,
    actionGroupState: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_ActionGroupSummary");
var de_Agent = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentArn: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString,
    agentName: import_smithy_client.expectString,
    agentResourceRoleArn: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString,
    clientToken: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    customerEncryptionKeyArn: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    failureReasons: import_smithy_client._json,
    foundationModel: import_smithy_client.expectString,
    idleSessionTTLInSeconds: import_smithy_client.expectInt32,
    instruction: import_smithy_client.expectString,
    preparedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    promptOverrideConfiguration: (_) => de_PromptOverrideConfiguration(_, context),
    recommendedActions: import_smithy_client._json,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_Agent");
var de_AgentActionGroup = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    actionGroupExecutor: (_) => (0, import_smithy_client._json)((0, import_core2.awsExpectUnion)(_)),
    actionGroupId: import_smithy_client.expectString,
    actionGroupName: import_smithy_client.expectString,
    actionGroupState: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString,
    apiSchema: (_) => (0, import_smithy_client._json)((0, import_core2.awsExpectUnion)(_)),
    clientToken: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    parentActionSignature: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentActionGroup");
var de_AgentAlias = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentAliasArn: import_smithy_client.expectString,
    agentAliasHistoryEvents: (_) => de_AgentAliasHistoryEvents(_, context),
    agentAliasId: import_smithy_client.expectString,
    agentAliasName: import_smithy_client.expectString,
    agentAliasStatus: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString,
    clientToken: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    routingConfiguration: import_smithy_client._json,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentAlias");
var de_AgentAliasHistoryEvent = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    endDate: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    routingConfiguration: import_smithy_client._json,
    startDate: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentAliasHistoryEvent");
var de_AgentAliasHistoryEvents = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_AgentAliasHistoryEvent(entry, context);
  });
  return retVal;
}, "de_AgentAliasHistoryEvents");
var de_AgentAliasSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_AgentAliasSummary(entry, context);
  });
  return retVal;
}, "de_AgentAliasSummaries");
var de_AgentAliasSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentAliasId: import_smithy_client.expectString,
    agentAliasName: import_smithy_client.expectString,
    agentAliasStatus: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    routingConfiguration: import_smithy_client._json,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentAliasSummary");
var de_AgentKnowledgeBase = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentId: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    knowledgeBaseState: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentKnowledgeBase");
var de_AgentKnowledgeBaseSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_AgentKnowledgeBaseSummary(entry, context);
  });
  return retVal;
}, "de_AgentKnowledgeBaseSummaries");
var de_AgentKnowledgeBaseSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    description: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    knowledgeBaseState: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentKnowledgeBaseSummary");
var de_AgentSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_AgentSummary(entry, context);
  });
  return retVal;
}, "de_AgentSummaries");
var de_AgentSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentId: import_smithy_client.expectString,
    agentName: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    latestAgentVersion: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentSummary");
var de_AgentVersion = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentArn: import_smithy_client.expectString,
    agentId: import_smithy_client.expectString,
    agentName: import_smithy_client.expectString,
    agentResourceRoleArn: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    customerEncryptionKeyArn: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    failureReasons: import_smithy_client._json,
    foundationModel: import_smithy_client.expectString,
    idleSessionTTLInSeconds: import_smithy_client.expectInt32,
    instruction: import_smithy_client.expectString,
    promptOverrideConfiguration: (_) => de_PromptOverrideConfiguration(_, context),
    recommendedActions: import_smithy_client._json,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    version: import_smithy_client.expectString
  });
}, "de_AgentVersion");
var de_AgentVersionSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_AgentVersionSummary(entry, context);
  });
  return retVal;
}, "de_AgentVersionSummaries");
var de_AgentVersionSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    agentName: import_smithy_client.expectString,
    agentStatus: import_smithy_client.expectString,
    agentVersion: import_smithy_client.expectString,
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_AgentVersionSummary");
var de_DataSource = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    dataSourceConfiguration: import_smithy_client._json,
    dataSourceId: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    name: import_smithy_client.expectString,
    serverSideEncryptionConfiguration: import_smithy_client._json,
    status: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    vectorIngestionConfiguration: import_smithy_client._json
  });
}, "de_DataSource");
var de_DataSourceSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_DataSourceSummary(entry, context);
  });
  return retVal;
}, "de_DataSourceSummaries");
var de_DataSourceSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    dataSourceId: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    name: import_smithy_client.expectString,
    status: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_DataSourceSummary");
var de_InferenceConfiguration = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    maximumLength: import_smithy_client.expectInt32,
    stopSequences: import_smithy_client._json,
    temperature: import_smithy_client.limitedParseFloat32,
    topK: import_smithy_client.expectInt32,
    topP: import_smithy_client.limitedParseFloat32
  });
}, "de_InferenceConfiguration");
var de_IngestionJob = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    dataSourceId: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    failureReasons: import_smithy_client._json,
    ingestionJobId: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    startedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    statistics: import_smithy_client._json,
    status: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_IngestionJob");
var de_IngestionJobSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_IngestionJobSummary(entry, context);
  });
  return retVal;
}, "de_IngestionJobSummaries");
var de_IngestionJobSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    dataSourceId: import_smithy_client.expectString,
    description: import_smithy_client.expectString,
    ingestionJobId: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    startedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    statistics: import_smithy_client._json,
    status: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_IngestionJobSummary");
var de_KnowledgeBase = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    createdAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_)),
    description: import_smithy_client.expectString,
    failureReasons: import_smithy_client._json,
    knowledgeBaseArn: import_smithy_client.expectString,
    knowledgeBaseConfiguration: import_smithy_client._json,
    knowledgeBaseId: import_smithy_client.expectString,
    name: import_smithy_client.expectString,
    roleArn: import_smithy_client.expectString,
    status: import_smithy_client.expectString,
    storageConfiguration: import_smithy_client._json,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_KnowledgeBase");
var de_KnowledgeBaseSummaries = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_KnowledgeBaseSummary(entry, context);
  });
  return retVal;
}, "de_KnowledgeBaseSummaries");
var de_KnowledgeBaseSummary = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    description: import_smithy_client.expectString,
    knowledgeBaseId: import_smithy_client.expectString,
    name: import_smithy_client.expectString,
    status: import_smithy_client.expectString,
    updatedAt: (_) => (0, import_smithy_client.expectNonNull)((0, import_smithy_client.parseRfc3339DateTimeWithOffset)(_))
  });
}, "de_KnowledgeBaseSummary");
var de_PromptConfiguration = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    basePromptTemplate: import_smithy_client.expectString,
    inferenceConfiguration: (_) => de_InferenceConfiguration(_, context),
    parserMode: import_smithy_client.expectString,
    promptCreationMode: import_smithy_client.expectString,
    promptState: import_smithy_client.expectString,
    promptType: import_smithy_client.expectString
  });
}, "de_PromptConfiguration");
var de_PromptConfigurations = /* @__PURE__ */ __name((output, context) => {
  const retVal = (output || []).filter((e) => e != null).map((entry) => {
    return de_PromptConfiguration(entry, context);
  });
  return retVal;
}, "de_PromptConfigurations");
var de_PromptOverrideConfiguration = /* @__PURE__ */ __name((output, context) => {
  return (0, import_smithy_client.take)(output, {
    overrideLambda: import_smithy_client.expectString,
    promptConfigurations: (_) => de_PromptConfigurations(_, context)
  });
}, "de_PromptOverrideConfiguration");
var deserializeMetadata = /* @__PURE__ */ __name((output) => ({
  httpStatusCode: output.statusCode,
  requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
  extendedRequestId: output.headers["x-amz-id-2"],
  cfId: output.headers["x-amz-cf-id"]
}), "deserializeMetadata");
var collectBodyString = /* @__PURE__ */ __name((streamBody, context) => (0, import_smithy_client.collectBody)(streamBody, context).then((body) => context.utf8Encoder(body)), "collectBodyString");
var _sRIUC = "skipResourceInUseCheck";
var _tK = "tagKeys";
var parseBody = /* @__PURE__ */ __name((streamBody, context) => collectBodyString(streamBody, context).then((encoded) => {
  if (encoded.length) {
    return JSON.parse(encoded);
  }
  return {};
}), "parseBody");
var parseErrorBody = /* @__PURE__ */ __name(async (errorBody, context) => {
  const value = await parseBody(errorBody, context);
  value.message = value.message ?? value.Message;
  return value;
}, "parseErrorBody");
var loadRestJsonErrorCode = /* @__PURE__ */ __name((output, data) => {
  const findKey = /* @__PURE__ */ __name((object, key) => Object.keys(object).find((k) => k.toLowerCase() === key.toLowerCase()), "findKey");
  const sanitizeErrorCode = /* @__PURE__ */ __name((rawValue) => {
    let cleanValue = rawValue;
    if (typeof cleanValue === "number") {
      cleanValue = cleanValue.toString();
    }
    if (cleanValue.indexOf(",") >= 0) {
      cleanValue = cleanValue.split(",")[0];
    }
    if (cleanValue.indexOf(":") >= 0) {
      cleanValue = cleanValue.split(":")[0];
    }
    if (cleanValue.indexOf("#") >= 0) {
      cleanValue = cleanValue.split("#")[1];
    }
    return cleanValue;
  }, "sanitizeErrorCode");
  const headerKey = findKey(output.headers, "x-amzn-errortype");
  if (headerKey !== void 0) {
    return sanitizeErrorCode(output.headers[headerKey]);
  }
  if (data.code !== void 0) {
    return sanitizeErrorCode(data.code);
  }
  if (data["__type"] !== void 0) {
    return sanitizeErrorCode(data["__type"]);
  }
}, "loadRestJsonErrorCode");

// src/commands/AssociateAgentKnowledgeBaseCommand.ts
var _AssociateAgentKnowledgeBaseCommand = class _AssociateAgentKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "AssociateAgentKnowledgeBase", {}).n("BedrockAgentClient", "AssociateAgentKnowledgeBaseCommand").f(void 0, void 0).ser(se_AssociateAgentKnowledgeBaseCommand).de(de_AssociateAgentKnowledgeBaseCommand).build() {
};
__name(_AssociateAgentKnowledgeBaseCommand, "AssociateAgentKnowledgeBaseCommand");
var AssociateAgentKnowledgeBaseCommand = _AssociateAgentKnowledgeBaseCommand;

// src/commands/CreateAgentActionGroupCommand.ts




var _CreateAgentActionGroupCommand = class _CreateAgentActionGroupCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "CreateAgentActionGroup", {}).n("BedrockAgentClient", "CreateAgentActionGroupCommand").f(CreateAgentActionGroupRequestFilterSensitiveLog, CreateAgentActionGroupResponseFilterSensitiveLog).ser(se_CreateAgentActionGroupCommand).de(de_CreateAgentActionGroupCommand).build() {
};
__name(_CreateAgentActionGroupCommand, "CreateAgentActionGroupCommand");
var CreateAgentActionGroupCommand = _CreateAgentActionGroupCommand;

// src/commands/CreateAgentAliasCommand.ts




var _CreateAgentAliasCommand = class _CreateAgentAliasCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "CreateAgentAlias", {}).n("BedrockAgentClient", "CreateAgentAliasCommand").f(void 0, void 0).ser(se_CreateAgentAliasCommand).de(de_CreateAgentAliasCommand).build() {
};
__name(_CreateAgentAliasCommand, "CreateAgentAliasCommand");
var CreateAgentAliasCommand = _CreateAgentAliasCommand;

// src/commands/CreateAgentCommand.ts




var _CreateAgentCommand = class _CreateAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "CreateAgent", {}).n("BedrockAgentClient", "CreateAgentCommand").f(CreateAgentRequestFilterSensitiveLog, CreateAgentResponseFilterSensitiveLog).ser(se_CreateAgentCommand).de(de_CreateAgentCommand).build() {
};
__name(_CreateAgentCommand, "CreateAgentCommand");
var CreateAgentCommand = _CreateAgentCommand;

// src/commands/CreateDataSourceCommand.ts




var _CreateDataSourceCommand = class _CreateDataSourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "CreateDataSource", {}).n("BedrockAgentClient", "CreateDataSourceCommand").f(void 0, void 0).ser(se_CreateDataSourceCommand).de(de_CreateDataSourceCommand).build() {
};
__name(_CreateDataSourceCommand, "CreateDataSourceCommand");
var CreateDataSourceCommand = _CreateDataSourceCommand;

// src/commands/CreateKnowledgeBaseCommand.ts




var _CreateKnowledgeBaseCommand = class _CreateKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "CreateKnowledgeBase", {}).n("BedrockAgentClient", "CreateKnowledgeBaseCommand").f(void 0, void 0).ser(se_CreateKnowledgeBaseCommand).de(de_CreateKnowledgeBaseCommand).build() {
};
__name(_CreateKnowledgeBaseCommand, "CreateKnowledgeBaseCommand");
var CreateKnowledgeBaseCommand = _CreateKnowledgeBaseCommand;

// src/commands/DeleteAgentActionGroupCommand.ts




var _DeleteAgentActionGroupCommand = class _DeleteAgentActionGroupCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteAgentActionGroup", {}).n("BedrockAgentClient", "DeleteAgentActionGroupCommand").f(void 0, void 0).ser(se_DeleteAgentActionGroupCommand).de(de_DeleteAgentActionGroupCommand).build() {
};
__name(_DeleteAgentActionGroupCommand, "DeleteAgentActionGroupCommand");
var DeleteAgentActionGroupCommand = _DeleteAgentActionGroupCommand;

// src/commands/DeleteAgentAliasCommand.ts




var _DeleteAgentAliasCommand = class _DeleteAgentAliasCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteAgentAlias", {}).n("BedrockAgentClient", "DeleteAgentAliasCommand").f(void 0, void 0).ser(se_DeleteAgentAliasCommand).de(de_DeleteAgentAliasCommand).build() {
};
__name(_DeleteAgentAliasCommand, "DeleteAgentAliasCommand");
var DeleteAgentAliasCommand = _DeleteAgentAliasCommand;

// src/commands/DeleteAgentCommand.ts




var _DeleteAgentCommand = class _DeleteAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteAgent", {}).n("BedrockAgentClient", "DeleteAgentCommand").f(void 0, void 0).ser(se_DeleteAgentCommand).de(de_DeleteAgentCommand).build() {
};
__name(_DeleteAgentCommand, "DeleteAgentCommand");
var DeleteAgentCommand = _DeleteAgentCommand;

// src/commands/DeleteAgentVersionCommand.ts




var _DeleteAgentVersionCommand = class _DeleteAgentVersionCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteAgentVersion", {}).n("BedrockAgentClient", "DeleteAgentVersionCommand").f(void 0, void 0).ser(se_DeleteAgentVersionCommand).de(de_DeleteAgentVersionCommand).build() {
};
__name(_DeleteAgentVersionCommand, "DeleteAgentVersionCommand");
var DeleteAgentVersionCommand = _DeleteAgentVersionCommand;

// src/commands/DeleteDataSourceCommand.ts




var _DeleteDataSourceCommand = class _DeleteDataSourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteDataSource", {}).n("BedrockAgentClient", "DeleteDataSourceCommand").f(void 0, void 0).ser(se_DeleteDataSourceCommand).de(de_DeleteDataSourceCommand).build() {
};
__name(_DeleteDataSourceCommand, "DeleteDataSourceCommand");
var DeleteDataSourceCommand = _DeleteDataSourceCommand;

// src/commands/DeleteKnowledgeBaseCommand.ts




var _DeleteKnowledgeBaseCommand = class _DeleteKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DeleteKnowledgeBase", {}).n("BedrockAgentClient", "DeleteKnowledgeBaseCommand").f(void 0, void 0).ser(se_DeleteKnowledgeBaseCommand).de(de_DeleteKnowledgeBaseCommand).build() {
};
__name(_DeleteKnowledgeBaseCommand, "DeleteKnowledgeBaseCommand");
var DeleteKnowledgeBaseCommand = _DeleteKnowledgeBaseCommand;

// src/commands/DisassociateAgentKnowledgeBaseCommand.ts




var _DisassociateAgentKnowledgeBaseCommand = class _DisassociateAgentKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "DisassociateAgentKnowledgeBase", {}).n("BedrockAgentClient", "DisassociateAgentKnowledgeBaseCommand").f(void 0, void 0).ser(se_DisassociateAgentKnowledgeBaseCommand).de(de_DisassociateAgentKnowledgeBaseCommand).build() {
};
__name(_DisassociateAgentKnowledgeBaseCommand, "DisassociateAgentKnowledgeBaseCommand");
var DisassociateAgentKnowledgeBaseCommand = _DisassociateAgentKnowledgeBaseCommand;

// src/commands/GetAgentActionGroupCommand.ts




var _GetAgentActionGroupCommand = class _GetAgentActionGroupCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetAgentActionGroup", {}).n("BedrockAgentClient", "GetAgentActionGroupCommand").f(void 0, GetAgentActionGroupResponseFilterSensitiveLog).ser(se_GetAgentActionGroupCommand).de(de_GetAgentActionGroupCommand).build() {
};
__name(_GetAgentActionGroupCommand, "GetAgentActionGroupCommand");
var GetAgentActionGroupCommand = _GetAgentActionGroupCommand;

// src/commands/GetAgentAliasCommand.ts




var _GetAgentAliasCommand = class _GetAgentAliasCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetAgentAlias", {}).n("BedrockAgentClient", "GetAgentAliasCommand").f(void 0, void 0).ser(se_GetAgentAliasCommand).de(de_GetAgentAliasCommand).build() {
};
__name(_GetAgentAliasCommand, "GetAgentAliasCommand");
var GetAgentAliasCommand = _GetAgentAliasCommand;

// src/commands/GetAgentCommand.ts




var _GetAgentCommand = class _GetAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetAgent", {}).n("BedrockAgentClient", "GetAgentCommand").f(void 0, GetAgentResponseFilterSensitiveLog).ser(se_GetAgentCommand).de(de_GetAgentCommand).build() {
};
__name(_GetAgentCommand, "GetAgentCommand");
var GetAgentCommand = _GetAgentCommand;

// src/commands/GetAgentKnowledgeBaseCommand.ts




var _GetAgentKnowledgeBaseCommand = class _GetAgentKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetAgentKnowledgeBase", {}).n("BedrockAgentClient", "GetAgentKnowledgeBaseCommand").f(void 0, void 0).ser(se_GetAgentKnowledgeBaseCommand).de(de_GetAgentKnowledgeBaseCommand).build() {
};
__name(_GetAgentKnowledgeBaseCommand, "GetAgentKnowledgeBaseCommand");
var GetAgentKnowledgeBaseCommand = _GetAgentKnowledgeBaseCommand;

// src/commands/GetAgentVersionCommand.ts




var _GetAgentVersionCommand = class _GetAgentVersionCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetAgentVersion", {}).n("BedrockAgentClient", "GetAgentVersionCommand").f(void 0, GetAgentVersionResponseFilterSensitiveLog).ser(se_GetAgentVersionCommand).de(de_GetAgentVersionCommand).build() {
};
__name(_GetAgentVersionCommand, "GetAgentVersionCommand");
var GetAgentVersionCommand = _GetAgentVersionCommand;

// src/commands/GetDataSourceCommand.ts




var _GetDataSourceCommand = class _GetDataSourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetDataSource", {}).n("BedrockAgentClient", "GetDataSourceCommand").f(void 0, void 0).ser(se_GetDataSourceCommand).de(de_GetDataSourceCommand).build() {
};
__name(_GetDataSourceCommand, "GetDataSourceCommand");
var GetDataSourceCommand = _GetDataSourceCommand;

// src/commands/GetIngestionJobCommand.ts




var _GetIngestionJobCommand = class _GetIngestionJobCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetIngestionJob", {}).n("BedrockAgentClient", "GetIngestionJobCommand").f(void 0, void 0).ser(se_GetIngestionJobCommand).de(de_GetIngestionJobCommand).build() {
};
__name(_GetIngestionJobCommand, "GetIngestionJobCommand");
var GetIngestionJobCommand = _GetIngestionJobCommand;

// src/commands/GetKnowledgeBaseCommand.ts




var _GetKnowledgeBaseCommand = class _GetKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "GetKnowledgeBase", {}).n("BedrockAgentClient", "GetKnowledgeBaseCommand").f(void 0, void 0).ser(se_GetKnowledgeBaseCommand).de(de_GetKnowledgeBaseCommand).build() {
};
__name(_GetKnowledgeBaseCommand, "GetKnowledgeBaseCommand");
var GetKnowledgeBaseCommand = _GetKnowledgeBaseCommand;

// src/commands/ListAgentActionGroupsCommand.ts




var _ListAgentActionGroupsCommand = class _ListAgentActionGroupsCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListAgentActionGroups", {}).n("BedrockAgentClient", "ListAgentActionGroupsCommand").f(void 0, void 0).ser(se_ListAgentActionGroupsCommand).de(de_ListAgentActionGroupsCommand).build() {
};
__name(_ListAgentActionGroupsCommand, "ListAgentActionGroupsCommand");
var ListAgentActionGroupsCommand = _ListAgentActionGroupsCommand;

// src/commands/ListAgentAliasesCommand.ts




var _ListAgentAliasesCommand = class _ListAgentAliasesCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListAgentAliases", {}).n("BedrockAgentClient", "ListAgentAliasesCommand").f(void 0, void 0).ser(se_ListAgentAliasesCommand).de(de_ListAgentAliasesCommand).build() {
};
__name(_ListAgentAliasesCommand, "ListAgentAliasesCommand");
var ListAgentAliasesCommand = _ListAgentAliasesCommand;

// src/commands/ListAgentKnowledgeBasesCommand.ts




var _ListAgentKnowledgeBasesCommand = class _ListAgentKnowledgeBasesCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListAgentKnowledgeBases", {}).n("BedrockAgentClient", "ListAgentKnowledgeBasesCommand").f(void 0, void 0).ser(se_ListAgentKnowledgeBasesCommand).de(de_ListAgentKnowledgeBasesCommand).build() {
};
__name(_ListAgentKnowledgeBasesCommand, "ListAgentKnowledgeBasesCommand");
var ListAgentKnowledgeBasesCommand = _ListAgentKnowledgeBasesCommand;

// src/commands/ListAgentsCommand.ts




var _ListAgentsCommand = class _ListAgentsCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListAgents", {}).n("BedrockAgentClient", "ListAgentsCommand").f(void 0, void 0).ser(se_ListAgentsCommand).de(de_ListAgentsCommand).build() {
};
__name(_ListAgentsCommand, "ListAgentsCommand");
var ListAgentsCommand = _ListAgentsCommand;

// src/commands/ListAgentVersionsCommand.ts




var _ListAgentVersionsCommand = class _ListAgentVersionsCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListAgentVersions", {}).n("BedrockAgentClient", "ListAgentVersionsCommand").f(void 0, void 0).ser(se_ListAgentVersionsCommand).de(de_ListAgentVersionsCommand).build() {
};
__name(_ListAgentVersionsCommand, "ListAgentVersionsCommand");
var ListAgentVersionsCommand = _ListAgentVersionsCommand;

// src/commands/ListDataSourcesCommand.ts




var _ListDataSourcesCommand = class _ListDataSourcesCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListDataSources", {}).n("BedrockAgentClient", "ListDataSourcesCommand").f(void 0, void 0).ser(se_ListDataSourcesCommand).de(de_ListDataSourcesCommand).build() {
};
__name(_ListDataSourcesCommand, "ListDataSourcesCommand");
var ListDataSourcesCommand = _ListDataSourcesCommand;

// src/commands/ListIngestionJobsCommand.ts




var _ListIngestionJobsCommand = class _ListIngestionJobsCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListIngestionJobs", {}).n("BedrockAgentClient", "ListIngestionJobsCommand").f(void 0, void 0).ser(se_ListIngestionJobsCommand).de(de_ListIngestionJobsCommand).build() {
};
__name(_ListIngestionJobsCommand, "ListIngestionJobsCommand");
var ListIngestionJobsCommand = _ListIngestionJobsCommand;

// src/commands/ListKnowledgeBasesCommand.ts




var _ListKnowledgeBasesCommand = class _ListKnowledgeBasesCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListKnowledgeBases", {}).n("BedrockAgentClient", "ListKnowledgeBasesCommand").f(void 0, void 0).ser(se_ListKnowledgeBasesCommand).de(de_ListKnowledgeBasesCommand).build() {
};
__name(_ListKnowledgeBasesCommand, "ListKnowledgeBasesCommand");
var ListKnowledgeBasesCommand = _ListKnowledgeBasesCommand;

// src/commands/ListTagsForResourceCommand.ts




var _ListTagsForResourceCommand = class _ListTagsForResourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "ListTagsForResource", {}).n("BedrockAgentClient", "ListTagsForResourceCommand").f(void 0, void 0).ser(se_ListTagsForResourceCommand).de(de_ListTagsForResourceCommand).build() {
};
__name(_ListTagsForResourceCommand, "ListTagsForResourceCommand");
var ListTagsForResourceCommand = _ListTagsForResourceCommand;

// src/commands/PrepareAgentCommand.ts




var _PrepareAgentCommand = class _PrepareAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "PrepareAgent", {}).n("BedrockAgentClient", "PrepareAgentCommand").f(void 0, void 0).ser(se_PrepareAgentCommand).de(de_PrepareAgentCommand).build() {
};
__name(_PrepareAgentCommand, "PrepareAgentCommand");
var PrepareAgentCommand = _PrepareAgentCommand;

// src/commands/StartIngestionJobCommand.ts




var _StartIngestionJobCommand = class _StartIngestionJobCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "StartIngestionJob", {}).n("BedrockAgentClient", "StartIngestionJobCommand").f(void 0, void 0).ser(se_StartIngestionJobCommand).de(de_StartIngestionJobCommand).build() {
};
__name(_StartIngestionJobCommand, "StartIngestionJobCommand");
var StartIngestionJobCommand = _StartIngestionJobCommand;

// src/commands/TagResourceCommand.ts




var _TagResourceCommand = class _TagResourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "TagResource", {}).n("BedrockAgentClient", "TagResourceCommand").f(void 0, void 0).ser(se_TagResourceCommand).de(de_TagResourceCommand).build() {
};
__name(_TagResourceCommand, "TagResourceCommand");
var TagResourceCommand = _TagResourceCommand;

// src/commands/UntagResourceCommand.ts




var _UntagResourceCommand = class _UntagResourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UntagResource", {}).n("BedrockAgentClient", "UntagResourceCommand").f(void 0, void 0).ser(se_UntagResourceCommand).de(de_UntagResourceCommand).build() {
};
__name(_UntagResourceCommand, "UntagResourceCommand");
var UntagResourceCommand = _UntagResourceCommand;

// src/commands/UpdateAgentActionGroupCommand.ts




var _UpdateAgentActionGroupCommand = class _UpdateAgentActionGroupCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateAgentActionGroup", {}).n("BedrockAgentClient", "UpdateAgentActionGroupCommand").f(UpdateAgentActionGroupRequestFilterSensitiveLog, UpdateAgentActionGroupResponseFilterSensitiveLog).ser(se_UpdateAgentActionGroupCommand).de(de_UpdateAgentActionGroupCommand).build() {
};
__name(_UpdateAgentActionGroupCommand, "UpdateAgentActionGroupCommand");
var UpdateAgentActionGroupCommand = _UpdateAgentActionGroupCommand;

// src/commands/UpdateAgentAliasCommand.ts




var _UpdateAgentAliasCommand = class _UpdateAgentAliasCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateAgentAlias", {}).n("BedrockAgentClient", "UpdateAgentAliasCommand").f(void 0, void 0).ser(se_UpdateAgentAliasCommand).de(de_UpdateAgentAliasCommand).build() {
};
__name(_UpdateAgentAliasCommand, "UpdateAgentAliasCommand");
var UpdateAgentAliasCommand = _UpdateAgentAliasCommand;

// src/commands/UpdateAgentCommand.ts




var _UpdateAgentCommand = class _UpdateAgentCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateAgent", {}).n("BedrockAgentClient", "UpdateAgentCommand").f(UpdateAgentRequestFilterSensitiveLog, UpdateAgentResponseFilterSensitiveLog).ser(se_UpdateAgentCommand).de(de_UpdateAgentCommand).build() {
};
__name(_UpdateAgentCommand, "UpdateAgentCommand");
var UpdateAgentCommand = _UpdateAgentCommand;

// src/commands/UpdateAgentKnowledgeBaseCommand.ts




var _UpdateAgentKnowledgeBaseCommand = class _UpdateAgentKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateAgentKnowledgeBase", {}).n("BedrockAgentClient", "UpdateAgentKnowledgeBaseCommand").f(void 0, void 0).ser(se_UpdateAgentKnowledgeBaseCommand).de(de_UpdateAgentKnowledgeBaseCommand).build() {
};
__name(_UpdateAgentKnowledgeBaseCommand, "UpdateAgentKnowledgeBaseCommand");
var UpdateAgentKnowledgeBaseCommand = _UpdateAgentKnowledgeBaseCommand;

// src/commands/UpdateDataSourceCommand.ts




var _UpdateDataSourceCommand = class _UpdateDataSourceCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateDataSource", {}).n("BedrockAgentClient", "UpdateDataSourceCommand").f(void 0, void 0).ser(se_UpdateDataSourceCommand).de(de_UpdateDataSourceCommand).build() {
};
__name(_UpdateDataSourceCommand, "UpdateDataSourceCommand");
var UpdateDataSourceCommand = _UpdateDataSourceCommand;

// src/commands/UpdateKnowledgeBaseCommand.ts




var _UpdateKnowledgeBaseCommand = class _UpdateKnowledgeBaseCommand extends import_smithy_client.Command.classBuilder().ep({
  ...commonParams
}).m(function(Command, cs, config, o) {
  return [
    (0, import_middleware_serde.getSerdePlugin)(config, this.serialize, this.deserialize),
    (0, import_middleware_endpoint.getEndpointPlugin)(config, Command.getEndpointParameterInstructions())
  ];
}).s("AmazonBedrockAgentBuildTimeLambda", "UpdateKnowledgeBase", {}).n("BedrockAgentClient", "UpdateKnowledgeBaseCommand").f(void 0, void 0).ser(se_UpdateKnowledgeBaseCommand).de(de_UpdateKnowledgeBaseCommand).build() {
};
__name(_UpdateKnowledgeBaseCommand, "UpdateKnowledgeBaseCommand");
var UpdateKnowledgeBaseCommand = _UpdateKnowledgeBaseCommand;

// src/BedrockAgent.ts
var commands = {
  AssociateAgentKnowledgeBaseCommand,
  CreateAgentCommand,
  CreateAgentActionGroupCommand,
  CreateAgentAliasCommand,
  CreateDataSourceCommand,
  CreateKnowledgeBaseCommand,
  DeleteAgentCommand,
  DeleteAgentActionGroupCommand,
  DeleteAgentAliasCommand,
  DeleteAgentVersionCommand,
  DeleteDataSourceCommand,
  DeleteKnowledgeBaseCommand,
  DisassociateAgentKnowledgeBaseCommand,
  GetAgentCommand,
  GetAgentActionGroupCommand,
  GetAgentAliasCommand,
  GetAgentKnowledgeBaseCommand,
  GetAgentVersionCommand,
  GetDataSourceCommand,
  GetIngestionJobCommand,
  GetKnowledgeBaseCommand,
  ListAgentActionGroupsCommand,
  ListAgentAliasesCommand,
  ListAgentKnowledgeBasesCommand,
  ListAgentsCommand,
  ListAgentVersionsCommand,
  ListDataSourcesCommand,
  ListIngestionJobsCommand,
  ListKnowledgeBasesCommand,
  ListTagsForResourceCommand,
  PrepareAgentCommand,
  StartIngestionJobCommand,
  TagResourceCommand,
  UntagResourceCommand,
  UpdateAgentCommand,
  UpdateAgentActionGroupCommand,
  UpdateAgentAliasCommand,
  UpdateAgentKnowledgeBaseCommand,
  UpdateDataSourceCommand,
  UpdateKnowledgeBaseCommand
};
var _BedrockAgent = class _BedrockAgent extends BedrockAgentClient {
};
__name(_BedrockAgent, "BedrockAgent");
var BedrockAgent = _BedrockAgent;
(0, import_smithy_client.createAggregatedClient)(commands, BedrockAgent);

// src/pagination/ListAgentActionGroupsPaginator.ts

var paginateListAgentActionGroups = (0, import_core.createPaginator)(BedrockAgentClient, ListAgentActionGroupsCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListAgentAliasesPaginator.ts

var paginateListAgentAliases = (0, import_core.createPaginator)(BedrockAgentClient, ListAgentAliasesCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListAgentKnowledgeBasesPaginator.ts

var paginateListAgentKnowledgeBases = (0, import_core.createPaginator)(BedrockAgentClient, ListAgentKnowledgeBasesCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListAgentVersionsPaginator.ts

var paginateListAgentVersions = (0, import_core.createPaginator)(BedrockAgentClient, ListAgentVersionsCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListAgentsPaginator.ts

var paginateListAgents = (0, import_core.createPaginator)(BedrockAgentClient, ListAgentsCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListDataSourcesPaginator.ts

var paginateListDataSources = (0, import_core.createPaginator)(BedrockAgentClient, ListDataSourcesCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListIngestionJobsPaginator.ts

var paginateListIngestionJobs = (0, import_core.createPaginator)(BedrockAgentClient, ListIngestionJobsCommand, "nextToken", "nextToken", "maxResults");

// src/pagination/ListKnowledgeBasesPaginator.ts

var paginateListKnowledgeBases = (0, import_core.createPaginator)(BedrockAgentClient, ListKnowledgeBasesCommand, "nextToken", "nextToken", "maxResults");

// src/index.ts
var import_util_endpoints = require("@aws-sdk/util-endpoints");
// Annotate the CommonJS export names for ESM import in node:

0 && (module.exports = {
  BedrockAgentServiceException,
  __Client,
  BedrockAgentClient,
  BedrockAgent,
  $Command,
  AssociateAgentKnowledgeBaseCommand,
  CreateAgentActionGroupCommand,
  CreateAgentAliasCommand,
  CreateAgentCommand,
  CreateDataSourceCommand,
  CreateKnowledgeBaseCommand,
  DeleteAgentActionGroupCommand,
  DeleteAgentAliasCommand,
  DeleteAgentCommand,
  DeleteAgentVersionCommand,
  DeleteDataSourceCommand,
  DeleteKnowledgeBaseCommand,
  DisassociateAgentKnowledgeBaseCommand,
  GetAgentActionGroupCommand,
  GetAgentAliasCommand,
  GetAgentCommand,
  GetAgentKnowledgeBaseCommand,
  GetAgentVersionCommand,
  GetDataSourceCommand,
  GetIngestionJobCommand,
  GetKnowledgeBaseCommand,
  ListAgentActionGroupsCommand,
  ListAgentAliasesCommand,
  ListAgentKnowledgeBasesCommand,
  ListAgentVersionsCommand,
  ListAgentsCommand,
  ListDataSourcesCommand,
  ListIngestionJobsCommand,
  ListKnowledgeBasesCommand,
  ListTagsForResourceCommand,
  PrepareAgentCommand,
  StartIngestionJobCommand,
  TagResourceCommand,
  UntagResourceCommand,
  UpdateAgentActionGroupCommand,
  UpdateAgentAliasCommand,
  UpdateAgentCommand,
  UpdateAgentKnowledgeBaseCommand,
  UpdateDataSourceCommand,
  UpdateKnowledgeBaseCommand,
  paginateListAgentActionGroups,
  paginateListAgentAliases,
  paginateListAgentKnowledgeBases,
  paginateListAgentVersions,
  paginateListAgents,
  paginateListDataSources,
  paginateListIngestionJobs,
  paginateListKnowledgeBases,
  AccessDeniedException,
  ActionGroupExecutor,
  ConflictException,
  ActionGroupState,
  APISchema,
  ActionGroupSignature,
  InternalServerException,
  ResourceNotFoundException,
  ServiceQuotaExceededException,
  ThrottlingException,
  ValidationException,
  AgentStatus,
  CreationMode,
  PromptState,
  PromptType,
  AgentAliasStatus,
  KnowledgeBaseState,
  DataSourceType,
  ChunkingStrategy,
  DataSourceStatus,
  IngestionJobStatus,
  IngestionJobFilterAttribute,
  IngestionJobFilterOperator,
  IngestionJobSortByAttribute,
  SortOrder,
  KnowledgeBaseType,
  KnowledgeBaseStorageType,
  KnowledgeBaseStatus,
  APISchemaFilterSensitiveLog,
  CreateAgentActionGroupRequestFilterSensitiveLog,
  AgentActionGroupFilterSensitiveLog,
  CreateAgentActionGroupResponseFilterSensitiveLog,
  GetAgentActionGroupResponseFilterSensitiveLog,
  UpdateAgentActionGroupRequestFilterSensitiveLog,
  UpdateAgentActionGroupResponseFilterSensitiveLog,
  PromptOverrideConfigurationFilterSensitiveLog,
  AgentFilterSensitiveLog,
  CreateAgentRequestFilterSensitiveLog,
  CreateAgentResponseFilterSensitiveLog,
  GetAgentResponseFilterSensitiveLog,
  UpdateAgentRequestFilterSensitiveLog,
  UpdateAgentResponseFilterSensitiveLog,
  AgentVersionFilterSensitiveLog,
  GetAgentVersionResponseFilterSensitiveLog
});
