// @aws-amplify/graphql-transformer-interfaces/lib/transformer-model-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionFieldType = exports.MutationFieldType = exports.QueryFieldType = void 0;
var QueryFieldType;
(function (QueryFieldType) {
    QueryFieldType["GET"] = "GET";
    QueryFieldType["LIST"] = "LIST";
    QueryFieldType["SYNC"] = "SYNC";
})(QueryFieldType = exports.QueryFieldType || (exports.QueryFieldType = {}));
var MutationFieldType;
(function (MutationFieldType) {
    MutationFieldType["CREATE"] = "CREATE";
    MutationFieldType["UPDATE"] = "UPDATE";
    MutationFieldType["DELETE"] = "DELETE";
})(MutationFieldType = exports.MutationFieldType || (exports.MutationFieldType = {}));
var SubscriptionFieldType;
(function (SubscriptionFieldType) {
    SubscriptionFieldType["ON_CREATE"] = "ON_CREATE";
    SubscriptionFieldType["ON_DELETE"] = "ON_DELETE";
    SubscriptionFieldType["ON_UPDATE"] = "ON_UPDATE";
})(SubscriptionFieldType = exports.SubscriptionFieldType || (exports.SubscriptionFieldType = {}));
