// @aws-amplify/core/dist/esm/clients/index.mjs
export { getDnsSuffix } from './endpoints/getDnsSuffix.mjs';
export { fetchTransferHandler } from './handlers/fetch.mjs';
export { unauthenticatedHandler } from './handlers/unauthenticated.mjs';
export { authenticatedHandler } from './handlers/authenticated.mjs';
export { signRequest } from './middleware/signing/signer/signatureV4/signRequest.mjs';
export { presignUrl } from './middleware/signing/signer/signatureV4/presignUrl.mjs';
export { EMPTY_HASH as EMPTY_SHA256_HASH } from './middleware/signing/signer/signatureV4/constants.mjs';
export { getHashedPayload } from './middleware/signing/signer/signatureV4/utils/getHashedPayload.mjs';
export { extendedEncodeURIComponent } from './middleware/signing/utils/extendedEncodeURIComponent.mjs';
export { signingMiddleware } from './middleware/signing/middleware.mjs';
export { retryMiddleware } from './middleware/retry/middleware.mjs';
export { jitteredBackoff } from './middleware/retry/jitteredBackoff.mjs';
export { getRetryDecider } from './middleware/retry/defaultRetryDecider.mjs';
export { userAgentMiddleware } from './middleware/userAgent/middleware.mjs';
export { parseMetadata } from './serde/responseInfo.mjs';
export { parseJsonBody, parseJsonError } from './serde/json.mjs';
export { withMemoization } from './utils/memoization.mjs';
