// @aws-amplify/auth/dist/cjs/providers/cognito/apis/rememberDevice.js
'use strict';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.rememberDevice = void 0;
const CognitoIdentityProvider_1 = require("../utils/clients/CognitoIdentityProvider");
const core_1 = require("@aws-amplify/core");
const types_1 = require("../utils/types");
const utils_1 = require("@aws-amplify/core/internals/utils");
const utils_2 = require("../utils/clients/CognitoIdentityProvider/utils");
const tokenProvider_1 = require("../tokenProvider");
const utils_3 = require("../../../utils");
/**
 * Marks device as remembered while authenticated.
 *
 * @throws - {@link UpdateDeviceStatusException} - Cognito service errors thrown when
 * setting device status to remembered using an invalid device key.
 * @throws AuthTokenConfigException - Thrown when the token provider config is invalid.
 */
async function rememberDevice() {
    const authConfig = core_1.Amplify.getConfig().Auth?.Cognito;
    (0, utils_1.assertTokenProviderConfig)(authConfig);
    const { tokens } = await (0, core_1.fetchAuthSession)();
    (0, types_1.assertAuthTokens)(tokens);
    const deviceMetadata = await tokenProvider_1.tokenOrchestrator?.getDeviceMetadata();
    (0, types_1.assertDeviceMetadata)(deviceMetadata);
    await (0, CognitoIdentityProvider_1.updateDeviceStatus)({
        region: (0, utils_2.getRegion)(authConfig.userPoolId),
        userAgentValue: (0, utils_3.getAuthUserAgentValue)(utils_1.AuthAction.RememberDevice),
    }, {
        AccessToken: tokens.accessToken.toString(),
        DeviceKey: deviceMetadata.deviceKey,
        DeviceRememberedStatus: 'remembered',
    });
}
exports.rememberDevice = rememberDevice;
